{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genotype Filtering Workflow for Extended Pink Data (Susitna and BC)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stacks version 1.32"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simplified steps:\n",
    "\n",
    "    Initial Filtering: \n",
    "        Retain loci with genotype rate >= 80% and loci with > 0.05 MAF in at least one population. \n",
    "\n",
    "    Run Stacks Populations with new whitelist for:\n",
    "        Genepop, VCF, plink outputs. \n",
    "\n",
    "    HDplot: \n",
    "        Identify paralogous loci. \n",
    "\n",
    "    Second Filtering: \n",
    "        Retain loci with highest MAF at each tag, and that pass HWE >0.05 in at least 9 populations, 50% of pops. \n",
    "        Filter out individuals that are genotyped in < 80% of loci. \n",
    "\n",
    "    Third Filtering: \n",
    "        Filter Haplotypes and One SNP per tag data sets for paralogs, Haplotypes for individuals \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data inventory:\n",
    "We have nearly raw genotypes for 492 pink salmon individuals from 18 paired populations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Background \n",
    "We want to generate 2 different data sets, one that has One SNP per Tag and is filtered in a way that makes it comparable to the previous work done on these populations. It will be filtered for genotype rate, MAF, HWE and paralogs and individuals will be removed that are not genotyped at at least 80% of the loci. \n",
    "\n",
    "For the intial 14 populations we filtered this way: \n",
    "    \n",
    "    Run with blacklist of loci to ignore, based on early filtering [We used whitelists this time]\n",
    "    Only export loci genotyped in 80% of individuals from Stacks [We filter for genotype rate in R this time]\n",
    "    Keep SNP with the highest MAF at each tag\n",
    "    Keep SNPs with MAF >0.05 in at least 1 population\n",
    "    Keep SNPs in HWE in at least 7 (50%) of populations p>0.05 \n",
    "    Keep SNPs with Fis > -0.5 (to filter for paralogs) [We will filter for paralogs differently this time]\n",
    "    Exclude samples genotyped at less than 80% of loci\n",
    "\n",
    "The second data set will be a Haplotype data set that is generated from the list of tags that contain loci that were genotyped in at least 80% of the individuals. The Haplotypes will have the same individuals removed as the One SNP per Tag data set, and will be filtered for paralogs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Where are we? \n",
    "We just generated 29 genepop files from running Stacks with 29 whitelists, one for each of the whitelist runs. We want to combine all the genepop files from the runs into one to analyze the output data. Garrett wrote R code that combines them all and does some data exploration and filtering, but it requires that the genepop files each be in a specific format so that R can handle them as a table. He only had a couple of files so it was no trouble to edit them by hand, but with 29, I need a more hands-off solution. \n",
    "\n",
    "I wrote this script to open each of the folders that was created and to take the genepop file and perform a couple of quick edits then write the edited version to another txt file outside of the folders so they are all in the same place, easy to read into R. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "EditGenepopForR.py\n",
    "#This is a script to take the genepop files that are made with iterative runs of populations and edit them to be used with Garrett's R code for filtering. \n",
    "\n",
    "import io\n",
    "import re\n",
    "\n",
    "for i in range(1,30): #change this to the # of whitelists you have\n",
    "\t#this opens your output file and it creates the file if it doesnt exist\n",
    "\tf2 = open(\"whitelist_\" + str(i) + \".genepop\", \"w+\")\n",
    "\twith io.open(r'whitelist_' + str(i) + r'\\batch_4.genepop', 'r') as f1:\n",
    "\t\tfor line in f1:\n",
    "\t\t\tif 'Stacks' in line:\n",
    "\t\t\t\tcontinue #if line contains('Stacks') skip it\n",
    "\t\t\tif re.match(r'pop', line): #if the line contains pop, skip it\n",
    "\t\t\t\tcontinue\n",
    "\t\t\tif re.match(r'P', line): #if the line contains a sample, \n",
    "\t\t\t\tf2.write(line) #write it to the output file unchanged\n",
    "\t\t\telse: \n",
    "\t\t\t\tf2.write(\"\\t\" + re.sub(r',', \"\\t\", line))\n",
    "\t\t\t#if it is the second line, (no pops or population name or stacks) write a tab, then split the remainder by replacing the , with tabs\n",
    "\t\tf1.close()\n",
    "\t\tf2.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial Filtering in R; Generation of Filtered Whitelist "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Garrett wrote some R code that takes all the genepop files that were generated from populations and have been edited for formatting and applies some quick filters on the data to get another whitelist of loci. That whitelist will be used on a run through populations to get the vcf and plink and genepop files that we want to do more indepth filtering with. \n",
    "\n",
    "I edited his code to filter for loci genotype rate (retain loci that were genotyped at at least 80%) and keep loci that have a MAF >=0.05 in at least one population.\n",
    "\n",
    "This imports the fist 12 genepop files from the whitelists because all those after 12 were empty of genotypes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Initial Filtering of Pink Stacks Population Output\n",
    "###    combines whitelists, and filters on MAF and locus genotype rate to create a final whitelist\n",
    "###    Uses a per population MAF instead of global filter and a per locus genotype rate\n",
    "### Garrett McKinney and Carolyn Tarpey | October 2017\n",
    "### ---------------------------------------\n",
    "\n",
    "#This is R code originally written by Garrett McKinney to combine and filter the output of Stacks populations. \n",
    "#It requires a genepop file for each of the whitelists used to run the populations command in Stacks. \n",
    "#The genepop files should be stripped of the header line that Stacks puts in and the second line should start with a tab so that \n",
    "#R will load it in as a table and the loci names should be tab deliminated, and there should be no \"pop\" between the samples.\n",
    "\n",
    "#install.packages(\"vcfR\")\n",
    "#install.packages(\"stringr\")\n",
    "#install.packages(\"ggplot2\")\n",
    "#install.packages(\"lazyeval\")\n",
    "#install.packages(\"adegenet\")\n",
    "#install.packages(\"dplyr\")\n",
    "#install.packages(\"DBI\")\n",
    "\n",
    "#Pink data filtering\n",
    "library(ggplot2)\n",
    "library(vcfR)\n",
    "library(stringr)\n",
    "library(dplyr)\n",
    "library(gdata)\n",
    "\n",
    "#load genepop files\n",
    "genos1<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_1.genepop\",colClasses=\"factor\")\n",
    "genos1[1:5,1:5]\n",
    "dim(genos1)\n",
    "genos2<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_2.genepop\",colClasses=\"factor\")\n",
    "genos2[1:5,1:5]\n",
    "dim(genos2)\n",
    "genos3<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_3.genepop\",colClasses=\"factor\")\n",
    "genos3[1:5,1:5]\n",
    "dim(genos3)\n",
    "genos4<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_4.genepop\",colClasses=\"factor\")\n",
    "genos4[1:5,1:5]\n",
    "dim(genos4)\n",
    "genos5<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_5.genepop\",colClasses=\"factor\")\n",
    "genos5[1:5,1:5]\n",
    "dim(genos5)\n",
    "genos6<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_6.genepop\",colClasses=\"factor\")\n",
    "genos6[1:5,1:5]\n",
    "dim(genos6)\n",
    "genos7<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_7.genepop\",colClasses=\"factor\")\n",
    "genos7[1:5,1:5]\n",
    "dim(genos7)\n",
    "genos8<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_8.genepop\",colClasses=\"factor\")\n",
    "genos8[1:5,1:5]\n",
    "dim(genos8)\n",
    "genos9<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_9.genepop\",colClasses=\"factor\")\n",
    "genos9[1:5,1:5]\n",
    "dim(genos9)\n",
    "genos10<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_10.genepop\",colClasses=\"factor\")\n",
    "genos10[1:5,1:5]\n",
    "dim(genos10)\n",
    "genos11<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_11.genepop\",colClasses=\"factor\")\n",
    "genos11[1:5,1:5]\n",
    "dim(genos11)\n",
    "genos12<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/OriginalWhitelists/Whitelist_Genepops/whitelist_12.genepop\",colClasses=\"factor\")\n",
    "genos12[1:5,1:5]\n",
    "\n",
    "#use cbind to combine data into single dataset\n",
    "allGenos<-cbind(genos1,genos2,genos3,genos4,genos5,genos6,genos7,genos8,genos9,genos10,genos11,genos12)\n",
    "allGenos[1:5,1:5]\n",
    "dim(allGenos)\n",
    "\n",
    "#get unique tags from full dataset\n",
    "allLoci<-data.frame(str_split_fixed(colnames(allGenos),\"_\",2))\n",
    "colnames(allLoci)<-c(\"Tag\",\"SNP\")\n",
    "allLoci$Locus<-colnames(allGenos)\n",
    "length(allLoci$Locus)\n",
    "length(unique(allLoci$Tag))\n",
    "#write.table(allLoci,\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/RawLoci_unfiltered.txt\",quote=FALSE,row.names=FALSE)\n",
    "#write.table(allGenos,\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/RawGenos_unfiltered.txt\",quote=FALSE,row.names=FALSE)\n",
    "\n",
    "#get genotype rate per locus\n",
    "locusGenoRate<-apply(allGenos,2,function(x) 1-(sum(x==\"0000\")/dim(allGenos)[1]))\n",
    "locusGenoRate<-data.frame(keyName=names(locusGenoRate), value=locusGenoRate, row.names=NULL)\n",
    "colnames(locusGenoRate)<-c(\"Locus\",\"GenoRate\")\n",
    "dim(locusGenoRate)\n",
    "head(locusGenoRate)\n",
    "#write.table(locusGenoRate,\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/LocusGenoRate.txt\",quote=FALSE,row.names=FALSE)\n",
    "\n",
    "#filterLoci with >= 80% genotype rate\n",
    "filteredLoci<-locusGenoRate[locusGenoRate$GenoRate>=0.80,]\n",
    "dim(filteredLoci)\n",
    "head(filteredLoci)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to original data set\n",
    "To see which loci from our original 16,681 loci are included in the new data set at this point:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "head(loci_16681)\n",
    "\n",
    "#list of all the loci names in allGenos, no \"X\" in the name \n",
    "allGenos_loci_test <- gsub(\"X\",\"\",colnames(allGenos))\n",
    "\n",
    "#loci that are in 16681 and 110433 \n",
    "in_16681_and_110433 <- intersect (loci_16681, allGenos_loci_test)\n",
    "length (in_16681_and_110433)\n",
    "setdiff(loci_16681, allGenos_loci_test)\n",
    "\n",
    "\n",
    "#if we want to see what was eliminated here\n",
    "excludedLoci_rate<-locusGenoRate[locusGenoRate$GenoRate<=0.80,]\n",
    "dim(excludedLoci_rate)\n",
    "head(excludedLoci_rate)\n",
    "\n",
    "excludedLoci<- as.vector(excludedLoci_rate[,1])\n",
    "head(excludedLoci)\n",
    "class(excludedLoci)\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "#loci that are in 16681 and 76761 (80% locus genotype rate)\n",
    "#list of all the loci names in allGenos, no \"X\" in the name \n",
    "filteredLoci_test <- gsub(\"X\",\"\", filteredLoci$Locus)\n",
    "\n",
    "in_16681_and_76761 <- intersect (loci_16681, filteredLoci_test)\n",
    "length(in_16681_and_110433)\n",
    "setdiff(loci_16681, filteredLoci_test)\n",
    "\n",
    "#We want to know the number of loci at the 31485 unique tags\n",
    "allLoci<-data.frame(str_split_fixed(colnames(allGenos),\"_\",2))\n",
    "colnames(allLoci)<-c(\"Tag\",\"SNP\")\n",
    "allLoci$Locus<-colnames(allGenos)\n",
    "length(allLoci$Locus)\n",
    "head(allLoci)\n",
    "\n",
    "\n",
    "#length(unique(allLoci$Tag))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering for MAF\n",
    "Continuing with the filtering the data set for MAF, retain loci that were >0.05 in at least one population. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#get unique tags from 80% genoRate filtered dataset\n",
    "retainedLoci_80perc<-data.frame(str_split_fixed(filteredLoci$Locus,\"_\",2))\n",
    "colnames(retainedLoci_80perc)<-c(\"Tag\",\"SNP\")\n",
    "retainedLoci_80perc$Locus<-filteredLoci$Locus\n",
    "head(retainedLoci_80perc)\n",
    "length(unique(retainedLoci_80perc$Tag))\n",
    "\n",
    "#format dataset to remove X from locus names\n",
    "filteredLociIDs<-filteredLoci$Locus\n",
    "filteredLociIDs<-gsub(\"X\",\"\",filteredLociIDs)\n",
    "\n",
    "#filter dataset to include only filtered loci\n",
    "filteredGenos<-allGenos[,colnames(allGenos)%in%filteredLoci$Locus]\n",
    "dim(filteredGenos)\n",
    "\n",
    "allGenos[1:5,1:5]\n",
    "filteredGenos[1:5,1:5]\n",
    "dim(filteredGenos)\n",
    "write.table(filteredGenos,\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/FirstFiltering/filteredGenos_just80PCgenorate_76761.txt\",quote=FALSE,row.names=TRUE)\n",
    "\n",
    "\n",
    "\n",
    "#get minor allele frequency for loci\n",
    "calculateMAF<-function(genotypes){\n",
    "  genotypeList<-sort(unique(genotypes))\n",
    "  genotypeList<-genotypeList[genotypeList != \"0000\"]\n",
    "  allelesList1<-substr(genotypeList,1,2)\n",
    "  allelesList2<-substr(genotypeList,3,4)\n",
    "  allelesList<-unique(c(allelesList1,allelesList2))\n",
    "  allele1Counts<-sum(str_count(genotypes,allelesList[1]))\n",
    "  allele2Counts<-sum(str_count(genotypes,allelesList[2]))\n",
    "  if(length(allelesList)==1){\n",
    "    MAF=0\n",
    "  }else if(allele1Counts>=allele2Counts){\n",
    "    MAF<-allele2Counts/(allele1Counts+allele2Counts)\n",
    "  }else{\n",
    "    MAF<-allele1Counts/(allele1Counts+allele2Counts)\n",
    "  }\n",
    "  return(MAF)\n",
    "}\n",
    "\n",
    "#assign each sample to a population, delete the first column (duplicate sample name) and rename the new column Pop\n",
    "popmap<- read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/PopMap.txt\")\n",
    "filteredGenos<-cbind(popmap,filteredGenos)\n",
    "filteredGenos$V1<- NULL\n",
    "colnames(filteredGenos)[1]<-\"Pop\"\n",
    "filteredGenos[1:5,1:5]\n",
    "#create a dataframe for our MAF results\n",
    "npops<- length(unique(filteredGenos$Pop))\n",
    "nloci<- length(filteredGenos[,-1])\n",
    "\n",
    "filteredGenos_POP_MAF<-matrix(nrow=npops, ncol=nloci) \n",
    "rownames(filteredGenos_POP_MAF)<-as.vector(unique(filteredGenos$Pop)) #name the rows by the population names\n",
    "colnames(filteredGenos_POP_MAF)<-colnames(filteredGenos[,-1])\n",
    "\n",
    "for (i in 1:length(unique(filteredGenos$Pop))){\n",
    "  tempPop<-(unique(filteredGenos$Pop)[i])\n",
    "  tempGeno<-subset(filteredGenos, Pop == tempPop)\n",
    "  #print(tempPop)\n",
    "  #print(tempGeno[1:5,1:5])\n",
    "  tempMAF<-apply(tempGeno[,-1],2,calculateMAF)\n",
    "  #print(head(tempMAF))\n",
    "  filteredGenos_POP_MAF[i,]<-c(tempMAF)\n",
    "  #print(filteredGenos_POP_MAF[1:i,1:7])\n",
    "}\n",
    "\n",
    "filteredGenos_POP_MAF[1:5,1:5]\n",
    "dim(filteredGenos_POP_MAF)\n",
    "#colnames(filteredGenos_POP_MAF[,1])\n",
    "\n",
    "##### Filter the loci by the MAF results by population\n",
    "##retain loci that were at least 0.05 in any of the 18 populations\n",
    "lociMAF<-vector()\n",
    "lociMAF<-apply(filteredGenos_POP_MAF,2,function(x) sum(x >=0.05, na.rm=TRUE))\n",
    "lociMAF<-data.frame(keynames=names(lociMAF), value=lociMAF, row.names = NULL)\n",
    "colnames(lociMAF)<-c(\"Locus\", \"PopsMAF\")\n",
    "\n",
    "#if the loci does not pass the test, delete it\n",
    "locipassedMAF<-vector()\n",
    "locipassedMAF<-subset(lociMAF, lociMAF[,2]!=0)\n",
    "dim(locipassedMAF)\n",
    "head(locipassedMAF)\n",
    "\n",
    "# a second copy to use later to see what matches with our 16681\n",
    "locipassedMAF_temp<-vector()\n",
    "locipassedMAF_temp<-subset(lociMAF, lociMAF[,2]!=0)\n",
    "head(locipassedMAF_temp)\n",
    "dim(locipassedMAF_temp)\n",
    "#locipassedMAF <-locipassedMAF_temp #reset\n",
    "\n",
    "#get unique tags from initial filtered dataset\n",
    "locipassedMAF<-data.frame(str_split_fixed(locipassedMAF$Locus,\"_\",2))\n",
    "colnames(locipassedMAF)<-c(\"Tag\",\"SNP\")\n",
    "\n",
    "## to figure out how many loci from the original data set will be genotyped at these tags\n",
    "UNIQUEtags31485<-unique(locipassedMAF$Tag)\n",
    "length(UNIQUEtags31485)\n",
    "\n",
    "#locipassedMAF_tags$Locus<-locipassedMAF$Locus\n",
    "head(locipassedMAF)\n",
    "# to get the unique tags for one SNP per tag\n",
    "locipassedMAF<-unique(locipassedMAF$Tag)\n",
    "length(locipassedMAF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to original data set\n",
    "To see how many loci of our original 16,681 SNPs are included in this stage of filtering of the new data set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#####################Test to see that what we got in our whitelist has the 16681 loci that we want\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "head(loci_16681)\n",
    "\n",
    "#list of all the loci names in locipassedMAF_temp, no \"X\" in the name \n",
    "#locipassedMAF_temp$Locus <- gsub(\"X\",\"\",locipassedMAF_temp$Locus)\n",
    "MAFlocilist <- gsub(\"X\",\"\",locipassedMAF_temp$Locus)\n",
    "length(MAFlocilist)\n",
    "\n",
    "#loci that are in 16662 and 48687 \n",
    "length(in_16681_and_110433)\n",
    "in_16662_and_48687 <- intersect(in_16681_and_110433, MAFlocilist)\n",
    "length(in_16662_and_48687)\n",
    "\n",
    "setdiff(in_16681_and_110433, MAFlocilist)\n",
    "\n",
    "\n",
    "### to figure out which loci will be included in the STACKS output files for the unique 31485 tags\n",
    "head(UNIQUEtags31485)\n",
    "allLoci[1:2,1:2]\n",
    "#UNIQUEtags31485 <- as.list(UNIQUEtags31485)\n",
    "loci4Tags <-allLoci[allLoci$Tag%in%UNIQUEtags31485,1:2]\n",
    "#filteredGenos<-allGenos[,colnames(allGenos)%in%filteredLoci$Locus]\n",
    "head(loci4Tags)\n",
    "length(loci4Tags$Tag)\n",
    "\n",
    "##Compare the 63758 loci that are in this set that is pulled from the 31485 loci with the 16681\n",
    "loci4Tags$Tag<- gsub(\"X\",\"\",loci4Tags$Tag) #remove the X in the tag name\n",
    "head(loci4Tags)\n",
    "loci4Tags$Locus<- paste(loci4Tags$Tag,loci4Tags$SNP, sep=\"_\") #this concatenates the two columns with the _ in the middle\n",
    "LociIN31485tags<- loci4Tags$Locus\n",
    "head(LociIN31485tags)\n",
    "in_16662_and_63758 <- intersect(in_16681_and_110433, LociIN31485tags)\n",
    "length(in_16662_and_63758)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export the whitelist of tags that contain loci retained to this point. \n",
    "This whitelist is to be used in Stacks to get a new Genepop file for the loci at these tags, a new haplotype file and to generate a VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#output list of filtered loci, need to open as binary file (using write binary, \"wb\") to give unix line endings\n",
    "#format dataset to remove X from locus names\n",
    "firstfilteredLociIDs<-gsub(\"X\",\"\",locipassedMAF)\n",
    "head(firstfilteredLociIDs)\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/firstfilteredLociIDs\", \"wb\")\n",
    "write.table(firstfilteredLociIDs,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Populations \n",
    "in Stacks was run with the new whitelist of 31,485 loci and the following command to generate a genepop file, and plink files for the loci at the 31,485 tags. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "populations -P ./ -b 4 -t 6 -M PopMap.txt -W firstFilteredLociIDs.txt --genepop --vcf --plink\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Though I included in the code, the VCF file is too large to generate with that method, the file was empty when I tried. instead Garrett gave me this code that he wrote that splits up the whitelist you give it into a set number of loci. I had ours split into lists of 10,000 loci, which generated 4 whitelists. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#generateWhitelist.pl\n",
    "#/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "my$whitelistCount=1;\n",
    "my$interval=10000;\n",
    "my$outfile=\"whitelist_\".$whitelistCount.\".txt\";\n",
    "open(OUTFILE,\">$outfile\")||die \"cannot open $outfile:$!\";\n",
    "my$count=0;\n",
    "while(my$line=<>){\n",
    "\tchomp $line;\n",
    "\t$count++;\n",
    "\tprint OUTFILE \"$line\\n\";\n",
    "\tif($count%$interval==0){\n",
    "\t\t#print OUTFILE \"$line\\n\";\n",
    "\t\tclose OUTFILE;\n",
    "\t\t$whitelistCount++;\n",
    "\t\t$outfile=\"whitelist_\".$whitelistCount.\".txt\";\n",
    "\t\topen(OUTFILE,\">$outfile\")||die \"cannot open $outfile:$!\";\n",
    "\t\t#$whitelistCount++;\n",
    "\t\t#print \"$line\\t$whitelistCount\\n\";\n",
    "\t}\n",
    "\t\n",
    "}\n",
    "close OUTFILE;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Then used his script called automateGenotyping_vcf.pl to run populations with those 4 whitelists. It takes the number of whitelists that you have and generates a line of code for the command line to run each of them in batch form. Below is an example of the code it generates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#automateGenotyping_vcf.pl\n",
    "#/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "foreach my$i (1..4){\n",
    "\tmy$command1=\"populations -b 4 -P ./ -M popMap.txt -W whitelist_\".$i.\".txt --vcf\";\n",
    "\tmy$command2=\"mkdir whitelist_\".$i;\n",
    "\tmy$command3=\"cp batch_4* whitelist_\".$i;\n",
    "\tprint \"$command1\\n\";\n",
    "\t`$command1`;\n",
    "\tprint \"$command2\\n\";\n",
    "\t`$command2`;\n",
    "\tprint \"$command3\\n\";\n",
    "\t`$command3`;\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "populations -b 4 -P ./ -M popMap.txt -W whitelist_1.txt --vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When Populations finished running for all four of these, I opened the files that it made and removed the header on the last three vcf files and then appended them one to the other. From there I made sure that they were accurately combined by testing the number of lines in the original files vs the combined file with the script that Garrett wrote.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#CountFileLines.py\n",
    "#This is a script to check the combined length of VCF files. \n",
    "#this runs with python, name of the script, VCF file to count the lines of \n",
    "\n",
    "import io\n",
    "import re\n",
    "import sys\n",
    "\n",
    "i = 0\n",
    "with open(sys.argv[1]) as f:\n",
    "\tfor i, l in enumerate(f):\n",
    "\t\tpass\n",
    "\ti + 1\n",
    "\n",
    "print i\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We don't use the VCF file for the next several steps, it will be used to run HDPlot and identify paralogous loci. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering for Highest MAF, HWE and individual genotype rate. \n",
    "The following set of R code is called SecondPinkFiltering and it takes the genepop file that was generated in Stacks for the 31,485 tags. It has to be edited for R so that it can be imported as a table. Use the python script detailed above, or just remove the title line, replace th \",\" between the loci names with tabs and start that line with a tab. Remove the \"Pop\" demarkations beween populations and it should import as a table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Secondary Filtering of Pink Stacks Population Output\n",
    "###   This code takes the final genepop file from Stacks and filters the genotypes: \n",
    "###    highest MAF SNP at each tag, HWE and filter for individuals\n",
    "###  Carolyn Tarpey | November 2017\n",
    "### ---------------------------------------\n",
    "\n",
    "#This is adapted from R code written by Garrett McKinney to combine and filter the output of Stacks populations. \n",
    "#It requires a genepop file stripped of the header line, the second line should start with a tab so that \n",
    "#R will load it in as a table and the loci names should be tab deliminated, and there should be no \"pop\" between the samples. \n",
    "#The python script EditGenepopForR.py will do this formatting for a Genepop file that doesn't have one snp per line\n",
    "\n",
    "#install.packages(\"vcfR\")\n",
    "#install.packages(\"stringr\")\n",
    "#install.packages(\"ggplot2\")\n",
    "#install.packages(\"lazyeval\")\n",
    "#install.packages(\"adegenet\")\n",
    "#install.packages(\"dplyr\")\n",
    "#install.packages(\"DBI\")\n",
    "#install.packages(\"tibble\")\n",
    "\n",
    "#Pink data filtering\n",
    "library(ggplot2)\n",
    "library(vcfR)\n",
    "library(stringr)\n",
    "library(dplyr)\n",
    "library(gdata)\n",
    "\n",
    "#load genepop files\n",
    "all_newGenos<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/batch_4_31485TAGS_genepop_4R.txt\", colClasses=\"factor\")\n",
    "all_newGenos[1:5,1:5]\n",
    "dim(all_newGenos)\n",
    "\n",
    "\n",
    "#get minor allele frequency for loci- want to keep SNP with highest MAF at each tag\n",
    "\n",
    "calculateMAF<-function(genotypes){\n",
    "  genotypeList<-sort(unique(genotypes))\n",
    "  genotypeList<-genotypeList[genotypeList != \"0000\"]\n",
    "  allelesList1<-substr(genotypeList,1,2)\n",
    "  allelesList2<-substr(genotypeList,3,4)\n",
    "  allelesList<-unique(c(allelesList1,allelesList2))\n",
    "  allele1Counts<-sum(str_count(genotypes,allelesList[1]))\n",
    "  allele2Counts<-sum(str_count(genotypes,allelesList[2]))\n",
    "  if(length(allelesList)==1){\n",
    "    MAF=0\n",
    "  }else if(allele1Counts>=allele2Counts){\n",
    "    MAF<-allele2Counts/(allele1Counts+allele2Counts)\n",
    "  }else{\n",
    "    MAF<-allele1Counts/(allele1Counts+allele2Counts)\n",
    "  }\n",
    "  return(MAF)\n",
    "}\n",
    "\n",
    "\n",
    "allGenos_oneSNP<-apply(all_newGenos,2,calculateMAF)\n",
    "head(allGenos_oneSNP)\n",
    "\n",
    "\n",
    "#put the results in a dataframe\n",
    "allGenos_oneSNP_temp<-data.frame(value=allGenos_oneSNP,row.names=names(all_newGenos))\n",
    "colnames(allGenos_oneSNP_temp)<-c(\"MAF\")\n",
    "head(allGenos_oneSNP_temp)\n",
    "\n",
    "#concatenate the locus names to the results. \n",
    "allLoci<-colnames(all_newGenos)\n",
    "length(allLoci)\n",
    "allGenos_oneSNP_temp$Locus<-allLoci\n",
    "head(allGenos_oneSNP_temp)\n",
    "\n",
    "#split the tags and snp positions\n",
    "allGenos_oneSNP_temp_tags<-data.frame(str_split_fixed(allGenos_oneSNP_temp$Locus,\"_\",2))\n",
    "colnames(allGenos_oneSNP_temp_tags)<-c(\"Tag\",\"SNP\")\n",
    "head(allGenos_oneSNP_temp_tags)\n",
    "allGenos_oneSNP_temp$Tag<-allGenos_oneSNP_temp_tags$Tag\n",
    "allGenos_oneSNP_temp$SNP<-allGenos_oneSNP_temp_tags$SNP\n",
    "head(allGenos_oneSNP_temp)\n",
    "#write.table(allGenos_oneSNP_temp,\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/allGenos_oneSNP_temp.txt\",quote=FALSE,row.names=FALSE)\n",
    "\n",
    "#Retain the SNP with the highest MAF per tag\n",
    "# This one gives an output with the correct number of unique tags, and spot checked in excel\n",
    "oneMAF<- allGenos_oneSNP_temp %>% group_by(Tag) %>% slice(which.max(MAF))\n",
    "head(oneMAF)\n",
    "#write.table(oneMAF,\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Filtering/oneMAF.txt\",quote=FALSE,row.names=FALSE)\n",
    "oneMAF<-as.data.frame(oneMAF)\n",
    "head(oneMAF)\n",
    "dim(oneMAF)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to original data set\n",
    "After idetifying which is the SNP with the highest MAF per tag, check to see how many of the loci retained so far are in the original 16,681 dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#####################Test to see that what we got in One SNP per tag by MAF has the 16681 loci that we want\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "head(loci_16681)\n",
    "\n",
    "#list of all the loci names in oneMAF_temp\n",
    "oneMAF_temp<- gsub(\"X\",\"\", oneMAF$Locus)\n",
    "head(oneMAF_temp)\n",
    "\n",
    "#loci that are in 16662 and 31845 \n",
    "length(in_16681_and_110433)\n",
    "in_16662_and_31485 <- intersect(in_16681_and_110433, oneMAF_temp)\n",
    "length(in_16662_and_31485)\n",
    "\n",
    "diff_16662_31485 <- setdiff(in_16681_and_110433, oneMAF_temp)\n",
    "length(diff_16662_31485)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the data set to include only one SNP per tag and export a list of the SNPS that meet the requirements so far. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###########################################\n",
    "\n",
    "#filter dataset to include only one SNP per tag loci\n",
    "length(oneMAF$Locus)\n",
    "\n",
    "filteredGenos_oneSNP<-all_newGenos[,oneMAF$Locus]\n",
    "dim(filteredGenos_oneSNP)\n",
    "\n",
    "all_newGenos[1:5,1:5]\n",
    "filteredGenos_oneSNP[1:5,1:5]\n",
    "\n",
    "\n",
    "############# Write a txt file with the list of loci- use as whitelist or subset genepop file\n",
    "\n",
    "#subset genepop file (for use with subset_genepop_by_SNPs.pl):\n",
    "\n",
    "#format dataset to remove X from locus names\n",
    "oneMAF_loci<-gsub(\"X\",\"\",oneMAF$Locus)\n",
    "head(oneMAF_loci)\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/oneSNPpertag_subsetgenepop.txt\", \"wb\")\n",
    "write.table(oneMAF_loci,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the list of loci that meet the filtering requirements so far to subset the Genepop file that was generated from Stacks most recently, with the 31,485 tag whitelist. The easiest way to do this is using the two sets of perl scripts to edit genepop files. The first, genepop_print_snp_names_1_per_line.pl takes the genepop file and formats it so that the line of loci names is one locus per line. The second, subset_genepop_by_SNPs.pl, takes list of loci you want genotypes for  and the the genepop file and outputs a new genepop file with just the information for those loci. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#genepop_print_snp_names_1_per_line.pl\n",
    "\n",
    "open(IN1, \"<$ARGV[0]\") or die \"Error!!! reading $ARGV[0] for reading\";\n",
    "\n",
    "$z=0;\n",
    "while(<IN1>)\n",
    "{\n",
    "  $line=$_;\n",
    "  chomp($line);\n",
    "  if($line=~m/Stacks/){print(\"Title Line:\\n\");}\n",
    "  if($z eq 1)\n",
    "  {\n",
    "  @SNP_names=split(\",\",$line);\n",
    "  foreach $SNP (@SNP_names){print(\"$SNP\\n\");}\n",
    "  }\n",
    "  elsif($z>1){print(\"$line\\n\");}\n",
    "  \n",
    "  $z++;\n",
    "}\n",
    "close IN1;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#subset_genepop_by_SNPs.pl\n",
    "#Subset a specified set of SNPs and output appended genepop file\n",
    "\n",
    "#arg1 SNPlist each line the full name of a SNP from the header in the genepop file\n",
    "#arg2 genepop file\n",
    "\n",
    "#usage\n",
    "#perl subset_genepop_by_SNPs.pl SNP_names.txt 96_snps_2_1_genepop.txt > new_genepop_file.txt\n",
    "#need to change this to match your SNP name prefix i.e. Ots, MY, One, RAD\n",
    "\n",
    "print(\"Title Line:\\n\");\n",
    "\n",
    "open(SNPs, \"<$ARGV[0]\") or die \"Error!!! reading $ARGV[0] for reading\";\n",
    "@SNP_names=();\n",
    "while(<SNPs>)\n",
    "{\n",
    "  chomp($_);\n",
    "  $SNP=\"$_\";\n",
    "  push(@SNP_names, $SNP);\n",
    "}\n",
    "close SNPs;\n",
    "\n",
    "#print(\"@SNP_names\");\n",
    "\n",
    "\n",
    "open(IN, \"<$ARGV[1]\") or die \"Error!!! reading $ARGV[1] for reading\"; #opens genepop file\n",
    "$z=0;\n",
    "@SNP_index=();\n",
    "while(<IN>)\n",
    " {\n",
    "    chomp($_);\n",
    "    $line=$_;\n",
    "    foreach $SNP (@SNP_names)\n",
    "    {\n",
    "      if($line eq $SNP) \n",
    "      {\n",
    "      print(\"$line\\n\");     \n",
    "      push(@SNP_index,$z);\n",
    "      }      \n",
    "    }\n",
    "    if($line !~m/Title|Pop|pop|,|Stacks/){$z++;}\n",
    "  }   \n",
    "  close IN;  \n",
    "\n",
    "open(IN, \"<$ARGV[1]\") or die \"Error!!! reading $ARGV[1] for reading\"; #opens genepop file\n",
    "while(<IN>)\n",
    "  {\n",
    "  chomp($_);\n",
    "  $line=$_;\n",
    "    if($line=~m/Pop|pop/){print(\"Pop\\n\");}\n",
    "    if($line =~ m/,/)\n",
    "    {\n",
    "    $ind=$line;\n",
    "    @split_ind_line=split(\" \" , $ind);\n",
    "    print(\"$split_ind_line[0]\\t\");\n",
    "      foreach $index (@SNP_index){print(\"$split_ind_line[$index+1]\\t\");}\n",
    "    print(\"\\n\");\n",
    "    }\n",
    "  }\n",
    "  close IN;\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use the new Genepop file in Genepop to get the HWE for the new loci. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "##################### Run Genepop with the genepop file for HWE and Fis and import the results here for filtering based on those \n",
    "\n",
    "#### Take the original Genepop file that was formatted for import at the beginning of this code and make it on snp per line\n",
    "###  Then subset it with the list that was exported above, and run it in Genepop for HWE\n",
    "###  Take the output of that and use the perl script get_HWresults_from_genepop.pl to get the results from the INF output file\n",
    "\n",
    "HWE_table_Pre_Filtered <- read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/batch_4_31485LOCI_HWE_results.txt\" ,\n",
    "                                     stringsAsFactors = FALSE, row.names= 1)\n",
    "HWE_pops <-c(\"AMUR_10\", \"AMUR_11\", \"SUSIT_13\", \"HAYLY_09\", \"HAYLY_10\", \"KOPE_91\", \"KOPE_96\", \"KUSHI_06\", \"KUSHI_07\",\n",
    "             \"LAKEL_06\", \"LAKEL_07\", \"NOME_91\", \"NOME_94\", \"SNOH_03\", \"SNOH_96\", \"SUSIT_14\", \"TAUY_09\", \"TAUY_12\")\n",
    "\n",
    "colnames(HWE_table_Pre_Filtered)<-c(\"AMUR_10\", \"AMUR_11\", \"SUSIT_13\", \"HAYLY_09\", \"HAYLY_10\", \"KOPE_91\", \"KOPE_96\", \"KUSHI_06\", \"KUSHI_07\",\n",
    "                                    \"LAKEL_06\", \"LAKEL_07\", \"NOME_91\", \"NOME_94\", \"SNOH_03\", \"SNOH_96\", \"SUSIT_14\", \"TAUY_09\", \"TAUY_12\")\n",
    "head(HWE_table_Pre_Filtered)\n",
    "dim(HWE_table_Pre_Filtered)\n",
    "nrow(HWE_table_Pre_Filtered)\n",
    "\n",
    "##### Filter the loci by the HWE\n",
    "##retain loci that were at least 0.05 in at least 9 of the populations\n",
    "\n",
    "loci_HWE_blank_test<-vector()\n",
    "loci_HWE_blank_test<-apply(HWE_table,1,function(x) sum(x <=0.05, na.rm=TRUE)-sum(x ==\"-\", na.rm=TRUE))\n",
    "loci_HWE_blank_test<-data.frame(keynames=names(loci_HWE_blank_test), value=loci_HWE_blank_test, row.names = NULL)\n",
    "colnames(loci_HWE_blank_test)<-c(\"Locus\", \"failedHWE-blanks\")\n",
    "head(loci_HWE_blank_test)\n",
    "dim(loci_HWE_blank_test)\n",
    "\n",
    "#if the loci does not pass the test, delete it from this group\n",
    "locipassedHWE_test<-vector()\n",
    "locipassedHWE_test<-subset(loci_HWE_blank_test, loci_HWE_blank_test[,2]<=9)\n",
    "dim(locipassedHWE_test)\n",
    "head(locipassedHWE_test)\n",
    "\n",
    "# #if the loci does not pass the test put it into a list for later\n",
    "# lociFailedHWE_test<-vector()\n",
    "# lociFailedHWE_test<-subset(loci_HWE_blank_test, loci_HWE_blank_test[,2]>=9)\n",
    "# dim(lociFailedHWE_test)\n",
    "# head(lociFailedHWE_test)\n",
    "# outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/Loci_FAILED_HWE.txt\", \"wb\")\n",
    "# write.table(lociFailedHWE_test[,1],outputFile,quote=FALSE,row.names=TRUE,col.names=FALSE,eol=\"\\n\")\n",
    "# close(outputFile)\n",
    "\n",
    "# #Retain the Loci that Passed HWE\n",
    "# HWE_table_Post_Filtered <-  HWE_table_Pre_Filtered[row.names(HWE_table_Pre_Filtered)%in%locipassedHWE_test$Locus, ] #this subsets the matrix by the row names in the list\n",
    "# dim(HWE_table_Post_Filtered)\n",
    "# head(HWE_table_Post_Filtered)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to original data set\n",
    "After filtering the loci for HWE, check to see how many of the loci retained so far are in the original 16,681 dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#####################Test to see that what we got in One SNP per tag by MAF has the 16681 loci that we want\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "head(loci_16681)\n",
    "\n",
    "#list of all the loci names in oneMAF_temp\n",
    "locipassedHWE_test_comp<- locipassedHWE_test$Locus\n",
    "head(locipassedHWE_test_comp)\n",
    "length(locipassedHWE_test_comp)\n",
    "\n",
    "#loci that are in 14637 and 30088, the loci that passed HWE\n",
    "length(in_16662_and_31485)\n",
    "in_14637_and_30088 <- intersect(in_16662_and_31485, locipassedHWE_test_comp)\n",
    "length(in_14637_and_30088)\n",
    "\n",
    "diff_14637_and_30088 <- setdiff(in_16662_and_31485, locipassedHWE_test_comp)\n",
    "length(diff_14637_and_30088)\n",
    "\n",
    "###########################################\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the loci for HWE and then identify and filter for the individual genotype rate, remove individuals that were genotyped at fewer than 80% of the loci. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##################subsample the genotypes based on the Loci that passed HWE\n",
    "#all_newGenos <- temp_allnewGenos #reset\n",
    "\n",
    "#format genotype dataset to remove X from locus names\n",
    "#temp_allnewGenos <- all_newGenos\n",
    "colnames(all_newGenos)<-gsub(\"X\",\"\",colnames(all_newGenos))\n",
    "all_newGenos[1:5,1:5]\n",
    "\n",
    "#filter dataset to include only loci that passed HWE\n",
    "all_newGenos_HWE_filtered<-all_newGenos[,colnames(all_newGenos)%in%locipassedHWE_test$Locus]\n",
    "dim(all_newGenos_HWE_filtered)\n",
    "all_newGenos_HWE_filtered[1:5,1:5]\n",
    "\n",
    "\n",
    "####Write list of loci that passed all filters so far:\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/Loci_passed_HWE_filtered.txt\", \"wb\")\n",
    "write.table(colnames(all_newGenos_HWE_filtered),outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "#get genotype rate per sample for filtered loci\n",
    "all_newGenos_HWE_inds<-apply(all_newGenos_HWE_filtered,1,function(x) 1-(sum(x==\"0000\")/dim(all_newGenos_HWE_filtered)[2]))\n",
    "all_newGenos_HWE_inds<-data.frame(keyName=names(all_newGenos_HWE_inds), value=all_newGenos_HWE_inds, row.names=NULL)\n",
    "colnames(all_newGenos_HWE_inds)<-c(\"Sample\",\"GenoRate\")\n",
    "dim(all_newGenos_HWE_inds)\n",
    "head(all_newGenos_HWE_inds)\n",
    "\n",
    "#plot ranked genotype rate for samples with filtered loci\n",
    "all_newGenos_HWE_inds_ranked<-all_newGenos_HWE_inds[order(all_newGenos_HWE_inds$GenoRate),]\n",
    "all_newGenos_HWE_inds_ranked$rank<-seq(1,dim(all_newGenos_HWE_inds_ranked)[1],by=1)\n",
    "ggplot()+geom_point(data=all_newGenos_HWE_inds_ranked,aes(x=rank,y=GenoRate))+theme_bw() + \n",
    "  geom_hline(aes(yintercept=0.80),lty=\"dashed\")+ggtitle(\"Sample Genotype Rate One SNP per Tag; Line Shows 80% Genotype Rate\")\n",
    "#ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/SampleGenotypeRateofLociGenotypedin80PCSamples.pdf\")\n",
    "\n",
    "#Keep samples with >=80% genotype rate\n",
    "filteredSamples<-all_newGenos_HWE_inds[all_newGenos_HWE_inds$GenoRate>=0.80,]\n",
    "filteredGenos_filteredSamples<-all_newGenos_HWE_filtered[rownames(all_newGenos_HWE_filtered)%in%filteredSamples$Sample,]\n",
    "dim(filteredGenos_filteredSamples)\n",
    "\n",
    "#get genotype rate per locus for filtered sample dataset\n",
    "locusGenoRate_filteredSamples<-apply(filteredGenos_filteredSamples,2,function(x) 1-(sum(x==\"0000\")/dim(all_newGenos_HWE_inds)[1]))\n",
    "locusGenoRate_filteredSamples<-data.frame(keyName=names(locusGenoRate_filteredSamples), value=locusGenoRate_filteredSamples, row.names=NULL)\n",
    "colnames(locusGenoRate_filteredSamples)<-c(\"Locus\",\"GenoRate\")\n",
    "dim(locusGenoRate_filteredSamples)\n",
    "head(locusGenoRate_filteredSamples)\n",
    "\n",
    "####Write a table of the genotypes/individuals that have passed all the filters so far:\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/FilteredGenos_FilteredInds.txt\", \"wb\")\n",
    "write.table(filteredGenos_filteredSamples,outputFile,quote=FALSE,row.names=TRUE,col.names=TRUE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "####Write a list of the individuals that have passed all the filters so far:\n",
    "filteredSample_names <- row.names(filteredGenos_filteredSamples)\n",
    "filteredSample_names <- as.list(filteredSamples$Sample)\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/filteredSample_names.txt\", \"wb\")\n",
    "write.table(filteredSample_names,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "####Write a list of the individuals that have FAILED the filters so far:\n",
    "#samples with >=80% genotype rate\n",
    "FAILED_filteredSamples<-all_newGenos_HWE_inds[all_newGenos_HWE_inds$GenoRate<0.80,]\n",
    "dim(FAILED_filteredSamples)\n",
    "head(FAILED_filteredSamples)\n",
    "FAILED_filteredSample_names <-as.list(FAILED_filteredSamples$Sample)\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/FAILED_filteredSamples.txt\", \"wb\")\n",
    "write.table(FAILED_filteredSample_names,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Did removing individuals drastically change the locus genotype rate of any of the loci? Plot the genotype rate of the new data set to see if any of the loci should be filtered because there are fewer individuals. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "######## Should we re-filter the loci based on genotype rate now that there are indv missing? NO\n",
    "\n",
    "#plot ranked genotype rate for samples with filtered loci\n",
    "x_ranked<-locusGenoRate_filteredSamples[order(locusGenoRate_filteredSamples$GenoRate),]\n",
    "x_ranked$rank<-seq(1,dim(x_ranked)[1],by=1)\n",
    "ggplot()+geom_point(data=x_ranked,aes(x=rank,y=GenoRate))+ggtitle(\"Locus Genotype Rate of Filtered 465 Samples\")+theme_bw()\n",
    "#ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/LocusGenotypeRateofFiltered465Samples.pdf\")\n",
    "\n",
    "#Which loci are genotyped at less than 80% Genotype rate? \n",
    "x_80<-locusGenoRate_filteredSamples[locusGenoRate_filteredSamples$GenoRate<0.80,]\n",
    "x_80_sort <-x_80[order(x_80$GenoRate, decreasing = FALSE),]\n",
    "dim(x_80_sort)\n",
    "head(x_80_sort)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run HDPlot on the VCF File to identify paralogous loci in the dataset\n",
    "The VCF file should be located on the desktop, it does not load from a network or external drive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Running HDPlot on the pink extended pops\n",
    "###   This takes a VCF file and edits it down based on previous filtering \n",
    "###   Then runs Garrett's code for HDplot to indentify paralogs\n",
    "###  Garrett McKinney (and Carolyn Tarpey) | December 2017\n",
    "### ---------------------------------------\n",
    "\n",
    "#This includes R code written by Garrett McKinney to identify paralogs: HDPlot, the fast version\n",
    "#It requires a VCF file, the results can be filtered for loci that we are interested later\n",
    "\n",
    "#install.packages(\"vcfR\")\n",
    "\n",
    "#Pink data filtering\n",
    "library(ggplot2)\n",
    "library(vcfR)\n",
    "library(stringr)\n",
    "library(dplyr)\n",
    "library(gdata)\n",
    "\n",
    "#imput the combined VCF file, this will not load from the network drive- must be on the desktop\n",
    "vcfInput<-read.vcfR(\"C:/Users/Carolyn/Desktop/HDPLOT/batch_4_combined.vcf\")\n",
    "\n",
    "vcfInput@fix[1:5,1:5]\n",
    "vcfInput@gt[1:5,1:5]\n",
    "\n",
    "HDPlot<-function(vcfData){\n",
    "  #set up results table\n",
    "  HDplotTable<-as.data.frame(matrix(NA,nrow=dim(vcfData@gt)[1],ncol=9))\n",
    "  colnames(HDplotTable)<-c(\"Locus_ID\",\"depth_a\",\"depth_b\",\"ratio\",\"num_hets\",\"num_samples\",\"het_perc\",\"std\",\"z\")\n",
    "  #format allele reads from vcf data into matrix of comma separated values\n",
    "  genos<-apply(vcfData@gt[,2:dim(vcfData@gt)[2]], 2, function(x) str_split_fixed(x,\":\",3)[,1])\n",
    "  rownames(genos)<-vcfData@fix[,3]\n",
    "  reads<-apply(vcfData@gt[,2:dim(vcfData@gt)[2]], 2, function(x) str_split_fixed(x,\":\",4)[,3])\n",
    "  rownames(reads)<-vcfData@fix[,3] \n",
    "  #replace . with 0\n",
    "  reads<-gsub(\"\\\\.\",\"0\",reads)\n",
    "  alleleReads_1<-apply(reads,2,function(x) str_split_fixed(x,\",\",2)[,1])\n",
    "  alleleReads_2<-apply(reads,2,function(x) str_split_fixed(x,\",\",2)[,2])\n",
    "  #convert to numeric format\n",
    "  alleleReads_1<-apply(alleleReads_1,2, function(x) as.numeric(x))\n",
    "  alleleReads_2<-apply(alleleReads_2,2, function(x) as.numeric(x))\n",
    "  rownames(alleleReads_1)<-vcfData@fix[,3]\n",
    "  rownames(alleleReads_2)<-vcfData@fix[,3]\n",
    "  #subset to heterozygous genotypes\n",
    "  #make genotype matrix where heterozygotes are 1 and other genotypes are 0\n",
    "  hetMatrix<-genos\n",
    "  hetMatrix<-apply(hetMatrix,2,function(x) dplyr::recode(x,'0/0'=0,'1/1'=0,'./.'=0,'0/1'=1,'1/0'=1))\n",
    "  #multiply read count matrices by heterozygote matrix to get read counts for heterozygotes\n",
    "  alleleReads_1_het<-alleleReads_1*hetMatrix\n",
    "  alleleReads_2_het<-alleleReads_2*hetMatrix\n",
    "  #rows are loci and columns are samples\n",
    "  #sum reads per allele per locus for heterozygous samples\n",
    "  A_reads<-apply(alleleReads_1_het,1,sum)\n",
    "  B_reads<-apply(alleleReads_2_het,1,sum)\n",
    "  totalReads<-A_reads+B_reads\n",
    "  ratio<-A_reads/totalReads\n",
    "  std<-sqrt(totalReads*0.5*0.5)\n",
    "  z<- -(totalReads/2-A_reads)/std\n",
    "  #get percent heterozygosity for each locus\n",
    "  numHets<-apply(hetMatrix,1,sum)\n",
    "  hetPerc<-numHets/dim(hetMatrix)[2]\n",
    "  \n",
    "  #assign results to HDplotTable\n",
    "  HDplotTable$Locus_ID<-vcfData@fix[,3]\n",
    "  HDplotTable$depth_a<-A_reads\n",
    "  HDplotTable$depth_b<-B_reads\n",
    "  HDplotTable$ratio<-ratio\n",
    "  HDplotTable$num_hets<-numHets\n",
    "  HDplotTable$num_samples<-dim(hetMatrix)[2]\n",
    "  HDplotTable$het_perc<-hetPerc\n",
    "  HDplotTable$std<-std\n",
    "  HDplotTable$z<-z\n",
    "  return(HDplotTable)\n",
    "}\n",
    "\n",
    "HDPlotResults<-HDPlot(vcfInput)\n",
    "dim(HDPlotResults)\n",
    "\n",
    "##plot the results of HDPlot in black and white\n",
    "ggplot()+geom_point(data=HDPlotResults,aes(x=het_perc,y=z),alpha=0.05)+theme_bw()\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/BlackWhiteHDPlotHet_per_Z.pdf\")\n",
    "\n",
    "ggplot()+geom_point(data=HDPlotResults,aes(x=het_perc,y=ratio),alpha=0.05)+theme_bw()\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/BlackWhiteHDPlotHet_per_Ratio.pdf\")\n",
    "\n",
    "#set thresholds for marker types\n",
    "thresh_H<-0.5\n",
    "thresh_H_divDup<-0.88\n",
    "thresh_D<-5\n",
    "thresh_Dneg<--5\n",
    "\n",
    "#Add paralog status to HDplot results table based on thresholds above\n",
    "paralogStatus<-function(data,thresh_H,thresh_D,thresh_Dneg,thresh_H_divDup){\n",
    "  #if(data$het_perc>thresh_H|data$z>thresh_D|data$z<thresh_Dneg){\n",
    "  H<-as.numeric(data[7])\n",
    "  D<-as.numeric(data[9])\n",
    "  if(is.na(D)){\n",
    "    paralog<-NA\n",
    "  }else if(H>=thresh_H_divDup){\n",
    "    paralog<-2\n",
    "  }else if(H>=thresh_H|D>thresh_D|D<thresh_Dneg){\n",
    "    paralog<-1\n",
    "  }else{\n",
    "    paralog<-0\n",
    "  }\n",
    "  return(paralog)\n",
    "}\n",
    "\n",
    "HDPlotResults$paralog<-apply(HDPlotResults,1,paralogStatus,thresh_H,thresh_D,thresh_Dneg,thresh_H_divDup)\n",
    "HDPlotResults$paralog<-as.factor(HDPlotResults$paralog)\n",
    "\n",
    "\n",
    "ggplot() + geom_point(data=HDPlotResults, aes(x=het_perc,y=z,color=paralog), alpha=0.05) + scale_color_manual(values=c(\"blue\",\"red\",\"dark green\"), \n",
    "  labels = c(\"Singletons\", \"Duplicates\", \"Diverged Duplicates\",\"\")) + theme_bw() + theme(legend.title=element_blank(),\n",
    "  legend.text = element_text(size= 15))+ guides(colour = guide_legend(override.aes = list(alpha = 1)))\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/ColorHDPlotHet_per_Z.pdf\") \n",
    "\n",
    "ggplot()+geom_point(data=HDPlotResults,aes(x=het_perc,y=ratio,color=paralog),alpha=0.05)+scale_color_manual(values=c(\"blue\",\"red\",\"dark green\"),\n",
    "  labels = c(\"Singletons\", \"Duplicates\", \"Diverged Duplicates\",\"\")) + theme_bw() + theme(legend.title=element_blank(),\n",
    "  legend.text = element_text(size= 15))+ guides(colour = guide_legend(override.aes = list(alpha = 1)))\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/ColorHDPlotHet_per_Ratio.pdf\")  \n",
    "\n",
    "#count the number of each type of locus                                                                                                                                                                                                                                 \n",
    "sum(HDPlotResults$paralog==0,na.rm=TRUE) # singletons\n",
    "sum(HDPlotResults$paralog==1,na.rm=TRUE) # duplicates\n",
    "sum(HDPlotResults$paralog==2,na.rm=TRUE) # diverged duplicates\n",
    "\n",
    "##convert the position number in the vcf file to the accurate position of the snp in the tag. \n",
    "head(HDPlotResults)\n",
    "vcfInput@fix[1:5,1:5]\n",
    "wrongPositions<-as.data.frame(vcfInput@fix)\n",
    "wrongPositions$POS<-as.numeric(as.character(wrongPositions$POS))\n",
    "correctPositions<-(wrongPositions$POS-2)%%94 \n",
    "correctPositions\n",
    "\n",
    "#put these positions in a data frame \n",
    "VCF_paralogs<-data.frame(value=correctPositions,row.names=names(correctPositions))\n",
    "colnames(VCF_paralogs)<-c(\"Position\")\n",
    "\n",
    "#concatenate the locus names to the results. \n",
    "wrongPositions$ID<-as.numeric(as.character(wrongPositions$ID))\n",
    "Tag <- wrongPositions$ID\n",
    "length(Tag)\n",
    "VCF_paralogs$Tag<-Tag\n",
    "head(VCF_paralogs)\n",
    "\n",
    "#add a column that is the full locus name \n",
    "VCF_paralogs$Locus <- paste(VCF_paralogs$Tag,VCF_paralogs$Position,sep=\"_\")\n",
    "\n",
    "\n",
    "#Add the paralog identity to the data frame\n",
    "VCF_paralogs$Identity<-HDPlotResults$paralog\n",
    "head(VCF_paralogs)\n",
    "\n",
    "#add an additional column with the identity in words\n",
    "VCF_paralogs$Identity_2<-as.character(HDPlotResults$paralog)\n",
    "VCF_paralogs$Identity_2[VCF_paralogs$Identity_2==0] <- \"singleton\"\n",
    "VCF_paralogs$Identity_2[VCF_paralogs$Identity_2==1] <- \"duplicate\"\n",
    "VCF_paralogs$Identity_2[VCF_paralogs$Identity_2==2] <- \"div_duplicate\"\n",
    "head(VCF_paralogs)\n",
    "\n",
    "#export a table of the loci and their paralog status\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/VCF_paralogs.txt\", \"wb\")\n",
    "write.table(VCF_paralogs,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "#sort the table and create lists of each of the loci for each paralog status\n",
    "singletons <- VCF_paralogs[VCF_paralogs$Identity==0,]\n",
    "singletons <- as.data.frame(singletons)\n",
    "singletons <- na.omit(singletons)\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/singletons.txt\", \"wb\")\n",
    "write.table(singletons,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "duplicates <- VCF_paralogs[VCF_paralogs$Identity==1,]\n",
    "duplicates <- as.data.frame(duplicates)\n",
    "duplicates <- na.omit(duplicates)\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/duplicates.txt\", \"wb\")\n",
    "write.table(duplicates,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "div_duplicates <- VCF_paralogs[VCF_paralogs$Identity==2,]\n",
    "div_duplicates <- as.data.frame(div_duplicates)\n",
    "div_duplicates <- na.omit(div_duplicates)\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/div_duplicates.txt\", \"wb\")\n",
    "write.table(div_duplicates,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "##Any paralog\n",
    "any_paralog <- VCF_paralogs[VCF_paralogs$Identity!=0,]\n",
    "any_paralog <- as.data.frame(any_paralog)\n",
    "any_paralog <- na.omit(any_paralog)\n",
    "dim(any_paralog)\n",
    "head(any_paralog)\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/any_paralog.txt\", \"wb\")\n",
    "write.table(any_paralog,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the results of HD Plot and use them to filter out paralogous loci in the Haplotype and One SNP per Tag datasets. In the following code, we will import the Haplotype file into R for the first time. It will be edited first. When we are filtering for loci that are paralogs, we remove any locus at a tag that contains a paralog. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Filter One SNP per tag dataset for paralogs\n",
    "###   Caculate the genotype rate of the Haplotype data set \n",
    "###   Remove the paralogs from Haplotype data set for Random Forest   \n",
    "###   Carolyn Tarpey | December 2017\n",
    "### ---------------------------------------\n",
    "\n",
    "#This code requires the lists of paralogs from HDPLOT_forPinks.R\n",
    "#The one SNP per tag genotypes should be filtered through SecondPinkFiltering.R already\n",
    "#The haplotype file is from STACKS using the 31,485 whitelist after initialPinkFiltering.R\n",
    "\n",
    "#install.packages(\"vcfR\")\n",
    "\n",
    "#Pink data filtering\n",
    "library(ggplot2)\n",
    "library(vcfR)\n",
    "library(stringr)\n",
    "library(dplyr)\n",
    "library(gdata)\n",
    "\n",
    "#input the haplotype file, make sure all the headers have no spaces\n",
    "Haplotype_file<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/STACKS/Ustacks/Whitelist31485_populations/batch_4.haplotypes.tsv\", header=TRUE)\n",
    "Haplotype_file[1:5,1:5]\n",
    "\n",
    "#input the single snp filtered genotype file, make sure all the headers have no spaces\n",
    "genotype_file<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/FilteredGenos_FilteredInds.txt\", header=TRUE, colClasses=\"factor\")\n",
    "genotype_file[1:5,1:5]\n",
    "dim(genotype_file)\n",
    "\n",
    "#load the list of loci that were any paralog from HDPLOT_forPinks.R  \n",
    "paralogs_alldata<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/any_paralog.txt\", colClasses=\"factor\")\n",
    "colnames(paralogs_alldata) <- c(\"Position\", \"Tag\", \"Locus\", \"Identity\", \"Identity_2\")\n",
    "head(paralogs_alldata)\n",
    "\n",
    "#load the list of loci that are singletons from HDPLOT_forPinks.R  \n",
    "singletons<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/singletons.txt\", colClasses=\"factor\")\n",
    "colnames(singletons) <- c(\"Position\", \"Tag\", \"Locus\", \"Identity\", \"Identity_2\")\n",
    "head(singletons)\n",
    "\n",
    "\n",
    "#load the list of individuals that passed the 80% genotype rate from SecondPinkFiltering.R \n",
    "indvs_list<-read.table(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/SecondFiltering/filteredSample_names.txt\", colClasses=\"character\")\n",
    "#indvs_list <- gsub(\",\",\"\",paste(indvs_list)) #the names are factors so you can't gsub it w/o paste\n",
    "#indvs_list <- as.data.frame(indvs_list)\n",
    "colnames(indvs_list) <-\"Sample\"\n",
    "head(indvs_list)\n",
    "dim(indvs_list)\n",
    "\n",
    "###################### Get lists of the paralog statuses\n",
    "\n",
    "#get the unique tags for the paralogs  \n",
    "paralog_tags<- unique(paralogs_alldata$Tag)\n",
    "length(paralog_tags)\n",
    "\n",
    "#get the unique tags for the singletons  \n",
    "singletons_tags<- unique(singletons$Tag)\n",
    "length(singletons_tags)\n",
    "\n",
    "#get the tags for singletons that are not in the paralog list\n",
    "just_singletons_tags<- setdiff(singletons_tags, paralog_tags)\n",
    "length(just_singletons_tags)\n",
    "\n",
    "\n",
    "######################## HAPLOTYPE FILTER THE INDIVIDUALS BASED ON 80% genotype rate from SecondPinkFiltering.R\n",
    "\n",
    "haplo_indvfil <- Haplotype_file[,colnames(Haplotype_file)%in%indvs_list$Sample]\n",
    "dim(haplo_indvfil)\n",
    "haplo_indvfil[1:5,1:5]\n",
    "#Haplotype_file[1:5,1:5]\n",
    "\n",
    "#add the tag number back to the beginning of the indv filtered haplo file\n",
    "Catalog_Ids <- Haplotype_file$Catalog_ID\n",
    "haplo_indvfil <- cbind(Catalog_Ids, haplo_indvfil)\n",
    "haplo_indvfil[1:5,1:5]\n",
    "\n",
    "\n",
    "######################## HAPLOTYPE FILTER THE GENOTYPES BASED ON PARALOG RESULTS\n",
    "\n",
    "#filter to keep only singleton haplotypes\n",
    "haplo_indv_parafil <- haplo_indvfil[haplo_indvfil$Catalog_Ids%in%just_singletons_tags, ]\n",
    "haplo_indv_parafil[1:5,1:5]\n",
    "dim(haplo_indv_parafil)\n",
    "\n",
    "######################## HAPLOTYPE calculate and plot idv and loci genotype rate to see if there are any more that need to be removed \n",
    "\n",
    "#Do the calculation for the individual genotype rate \n",
    "\n",
    "#get genotype rate per individual\n",
    "IND_genorate<-apply(haplo_indv_parafil,2,function(x) 1-(sum(x==\"-\")/length(haplo_indv_parafil$Catalog_Ids)[1]))\n",
    "IND_genorate<-data.frame(keyName=names(IND_genorate), value=IND_genorate, row.names=NULL)\n",
    "IND_genorate <- IND_genorate[-1,] #for the catalogID \n",
    "colnames(IND_genorate)<-c(\"Sample\",\"GenoRate\")\n",
    "dim(IND_genorate)\n",
    "head(IND_genorate)\n",
    "min(IND_genorate[,2])\n",
    "\n",
    "#plot ranked genotype rate for samples with filtered loci\n",
    "inds_ranked <- IND_genorate[order(IND_genorate$GenoRate),]\n",
    "inds_ranked$rank<-seq(1,dim(inds_ranked)[1],by=1)\n",
    "ggplot()+geom_point(data=inds_ranked,aes(x=rank,y=GenoRate))+ggtitle(\"Individual Genotype Rate for filtered Haplotypes\") +theme_bw()\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/INDGenotypeRateForHaplotypesFiltered.pdf\")\n",
    "\n",
    "\n",
    "#Do the calculation for the haplotype genotype rate \n",
    "#get genotype rate per locus\n",
    "CAT_IDS <-haplo_indv_parafil$Catalog_Ids\n",
    "#head(CAT_IDS)\n",
    "locus_genorate<-apply(haplo_indv_parafil[,-1],1,function(x) 1-(sum(x==\"-\")/(ncol(haplo_indv_parafil)-1))) #remove one from row count for the CatalogID\n",
    "locus_genorate<-data.frame(keyName=CAT_IDS, value=locus_genorate)\n",
    "colnames(locus_genorate)<-c(\"Locus\",\"GenoRate\")\n",
    "head(locus_genorate)\n",
    "dim(locus_genorate)\n",
    "min(locus_genorate[,2])\n",
    "\n",
    "#plot ranked genotype rate for samples with filtered loci\n",
    "Loci_ranked <- locus_genorate[order(locus_genorate$GenoRate),]\n",
    "Loci_ranked$rank<-seq(1,dim(locus_genorate)[1],by=1)\n",
    "ggplot()+geom_point(data=Loci_ranked,aes(x=rank,y=GenoRate)) +ggtitle(\"Locus Genotype Rate for filtered Haplotypes\")+theme_bw()\n",
    "ggsave(filename=\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/LOCUSGenotypeRateForHaplotypesFiltered.pdf\")\n",
    "\n",
    "\n",
    "###Don't have to do any additional Genotype or Individual filtering\n",
    "###### export the haplotype file that has been filtered for the individuals and paralogs removed\n",
    "\n",
    "outputFile<-file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/HDPlot/Haplotypes_with465ins_singletons.txt\", \"wb\")\n",
    "write.table(haplo_indv_parafil,outputFile,quote=FALSE,row.names=FALSE,col.names=TRUE,eol=\"\\n\")\n",
    "close(outputFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter the One SNP per Tag dataset to retain loci that are not paralogs. We filter out any locus that was considered anything but a singleton. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### GENOTYPE filter the loci for paralogs\n",
    "######################## FILTER THE GENOTYPES BASED ON PARALOG RESULTS\n",
    "\n",
    "#format dataset to remove X from locus names\n",
    "genotype_file_t<-gsub(\"X\",\"\",colnames(genotype_file))\n",
    "head(genotype_file_t)\n",
    "length(genotype_file_t)\n",
    "colnames(genotype_file) <- genotype_file_t\n",
    "genotype_file[1:5,1:5]\n",
    "\n",
    "#pull the loci names from the genotype file \n",
    "genotype_loci <- colnames(genotype_file)\n",
    "head(genotype_loci)\n",
    "\n",
    "#break up the loci names into their components in a table \n",
    "loci_table_t <- as.data.frame(str_split_fixed(genotype_loci, \"_\", 2))\n",
    "ncol(loci_table_t)\n",
    "colnames(loci_table_t) <- c(\"Tag\", \"Pos\")\n",
    "loci_table_t$Locus <- genotype_loci\n",
    "head(loci_table_t)\n",
    "dim(loci_table_t)\n",
    "\n",
    "#get the tags for singletons that are not in the paralog list\n",
    "just_singletons_tags <- as.numeric(setdiff(singletons_tags, paralog_tags))\n",
    "length(just_singletons_tags)\n",
    "head(just_singletons_tags)\n",
    "\n",
    "singletons_to_keep <- loci_table_t[loci_table_t$Tag%in%just_singletons_tags,]\n",
    "dim(singletons_to_keep)\n",
    "head(singletons_to_keep)\n",
    "\n",
    "singletons_locus_keep<- singletons_to_keep$Locus\n",
    "length(singletons_locus_keep)\n",
    "head(singletons_locus_keep)\n",
    "\n",
    "filtered_singleton_genotypes <- genotype_file[,colnames(genotype_file)%in%singletons_locus_keep]\n",
    "dim(filtered_singleton_genotypes)\n",
    "filtered_singleton_genotypes[1:5,1:5]\n",
    "\n",
    "####Write a table of the genotypes/individuals that have passed all the filters so far:\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/ThirdFiltering/Singleton_One_Tag_Genotypes.txt\", \"wb\")\n",
    "write.table(filtered_singleton_genotypes,outputFile,quote=FALSE,row.names=TRUE,col.names=TRUE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n",
    "####Write a list of the SNPSthat have passed all the filters so far:\n",
    "outputFile <- file(\"Z:/WORK/TARPEY/Exp_Pink_Pops/FilteringGenotypes/ThirdFiltering/Singleton_One_Tag_SNPLIST.txt\", \"wb\")\n",
    "write.table(singletons_locus_keep,outputFile,quote=FALSE,row.names=FALSE,col.names=FALSE,eol=\"\\n\")\n",
    "close(outputFile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison to original data set\n",
    "After filtering the One SNP per Tag loci and the Haplotypes for Paralogs, check to see how many of the loci retained so far are in the original 16,681 dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#####################Test to see that what we got in the final paralog filtered one snp per tag data set has the 16681 loci\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "head(loci_16681)\n",
    "\n",
    "#list of all the loci that were singletons that we kept\n",
    "length(singletons_locus_keep)\n",
    "\n",
    "#loci that are in  the singleton loci and the remaing 14629 loci of the 16681\n",
    "length(in_14637_and_30088)\n",
    "in_14637_and_23759 <- intersect(in_14637_and_30088, singletons_locus_keep)\n",
    "length(in_14637_and_23759)\n",
    "\n",
    "diff_14637_and_23759 <- setdiff(in_14637_and_30088, singletons_locus_keep)\n",
    "length(diff_14637_and_23759)\n",
    "\n",
    "#####################Test to see that what we got in the final paralog filtered Haplotype data set has the 16681 loci\n",
    "\n",
    "#if we want to see which of our 16681 snps is included at this point\n",
    "loci_16681 <- readLines(\"Z:/WORK/TARPEY/Pink_Populations/listof16681LOCI.txt\")\n",
    "loci_16681_tags <- as.data.frame(str_split_fixed(loci_16681, \"_\", 2))\n",
    "head(loci_16681_tags)\n",
    "colnames(loci_16681_tags)<- c(\"Tag\",\"Pos\")\n",
    "Just_loci_16681_tags<-loci_16681_tags$Tag \n",
    "length(Just_loci_16681_tags)\n",
    "\n",
    "#list of all the loci that were singletons that we kept\n",
    "length(just_singletons_tags)\n",
    "\n",
    "#loci that are in  the singleton loci and the remaing 14629 loci of the 16681\n",
    "\n",
    "HaplotypeTags_16681Tags <- intersect(just_singletons_tags, Just_loci_16681_tags)\n",
    "length(HaplotypeTags_16681Tags)\n",
    "\n",
    "diff_14637_and_23759 <- setdiff(in_14637_and_30088, singletons_locus_keep)\n",
    "length(diff_14637_and_23759)\n",
    "\n",
    "###########################################\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### At this point we have two filtered data sets!\n",
    "One is the Haplotypes, and the other are loci with One SNP per Tag. They are ready for data analysis. We will start by plotting some of the metrics for each of the factors we filtered on for each population, to see the change in the population for that metric before and after filtering. That is found in the R code PlotPopStats.R.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
