{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter genotype data\n",
    "This notebook was adapted from one written by Ryan Waples: \n",
    "https://github.com/rwaples/chum_populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bring:\n",
    "* raw genotypes in [MAP/PED format](http://pngu.mgh.harvard.edu/~purcell/plink/data.shtml)\n",
    "    individual pedigree information and genotype calls\n",
    "    * \\*.ped file individual pedigree information and genotype calls\n",
    "    * \\*.map file - variant information file\n",
    "   \n",
    "These files can be generated by the **`populations`** program within **`Stacks`** using the \"--plink\" flag\n",
    "\n",
    "### Take away:\n",
    "* Set of genotypes passing filtering criteria.\n",
    "\n",
    "### Programs used:\n",
    "* [PLINK v1.9b](http://pngu.mgh.harvard.edu/~purcell/plink2/index.html)\n",
    "    \n",
    "### Steps\n",
    "1. Setup\n",
    "2. Filter based on missingness.\n",
    "3. Test for Hardy-Weinberg equilibrium (HWE) and quantify minor allele frequency (MAF) within each population.\n",
    "4. Apply filters based on HWE and MAF.\n",
    "5. Keep only one SNP per locus.\n",
    "6. Summary statistics\n",
    "\n",
    "Im not sure how Ryan ran all the above steps, I didn't do them here in this notebook. I used the command line. For the Python code, I ran them as a python program, for the plink commands I either copied them into the command line or I ran them as a batchfile .bat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1\n",
    "### Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Change to the directory with your map/ped files. I copied the program here too, I have something else that also runs with the commn  The filtering process will generate many intermediate files, so I made a separate directory for them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cd G:\\Analysis\\Pop_analysis\\RenamedCatalog2\\FilteringDips\\plink_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### universal plink info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "raw_genotypes  = \"batch_3.plink\"\n",
    "univseral_plink_commands = \"--allow-extra-chr --allow-no-sex --write-snplist\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### filtering parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pops_to_keep = ['1','2','3','4','5','6','7','8','9','10','11','12','13','14']\n",
    "\n",
    "# missingness\n",
    "locus_max_miss = 0.25\n",
    "ind_max_miss = 0.25\n",
    "\n",
    "# MAF\n",
    "local_MAF_threshold = 0.05\n",
    "global_MAF_threshold = 0.05\n",
    "\n",
    "# HWE\n",
    "local_HWE_pval = .05\n",
    "HWE_pop_threshold = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This command imports the raw genotypes:\n",
    "\n",
    "--keep-cluster-names can be used individually or in combination to define a list of clusters to keep; all samples not in one of those clusters are then removed from the current analysis. --keep-cluster-names takes a space-delimited sequence of cluster names on the command line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --file batch_3.plink --allow-extra-chr --allow-no-sex --write-snplist --family --keep-cluster-names 1 2 3 4 5 6 7 8 9 10 11 12 13 14 --make-bed --out filter_pops\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Step 2 \n",
    "###filter based on missingness"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### remove samples not genotyped in at least 0.25% of loci "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile filter_pops --mind 0.25 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out filter_inds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### remove samples not genotyped in at least 0.25% of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile  filter_inds --geno 0.25 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out miss_final"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 3\n",
    "### HWE and MAF within each population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a separate file for each population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I ran this as a python script called pops_to_keep.py to give me all the plink commands I would need to make a batch file, instead of writing them out for each of the 14 populations by hand: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pops_to_keep = ['1','2','3','4','5','6','7','8','9','10','11','12','13','14']\n",
    "\n",
    "#Create a separate file for each population\n",
    "for x in pops_to_keep:\n",
    "\tprint \"plink --bfile  miss_final --family --keep-cluster-names\", x , \"--allow-extra-chr --allow-no-sex --write-snplist --autosome-num 50 --make-bed --out fam_\", x\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Create a separate file for each population\n",
    "#ran this as the plink.bat file, it has the windows line endings\n",
    "\n",
    "plink --bfile  miss_final --family --keep-cluster-names 1 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_1\n",
    "plink --bfile  miss_final --family --keep-cluster-names 2 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_2\n",
    "plink --bfile  miss_final --family --keep-cluster-names 3 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_3\n",
    "plink --bfile  miss_final --family --keep-cluster-names 4 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_4\n",
    "plink --bfile  miss_final --family --keep-cluster-names 5 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_5\n",
    "plink --bfile  miss_final --family --keep-cluster-names 6 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_6\n",
    "plink --bfile  miss_final --family --keep-cluster-names 7 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_7\n",
    "plink --bfile  miss_final --family --keep-cluster-names 8 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_8\n",
    "plink --bfile  miss_final --family --keep-cluster-names 9 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_9\n",
    "plink --bfile  miss_final --family --keep-cluster-names 10 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_10\n",
    "plink --bfile  miss_final --family --keep-cluster-names 11 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_11\n",
    "plink --bfile  miss_final --family --keep-cluster-names 12 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_12\n",
    "plink --bfile  miss_final --family --keep-cluster-names 13 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_13\n",
    "plink --bfile  miss_final --family --keep-cluster-names 14 --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out fam_14"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculate HWE statistics, write a list of snps passing HWE in each pop and calculate MAF, make a list of snps with MAF > threshold in each pop\n",
    "this python program was also run to make the commands for a batchfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pops_to_keep = ['1','2','3','4','5','6','7','8','9','10','11','12','13','14']\n",
    "\n",
    "#Calculate HWE statistics, write a list of snps passing HWE in each pop\n",
    "for x in pops_to_keep:\n",
    "    print\"plink --bfile  fam_\",x, \"--hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist --autosome-num 50 --out fam_\",x,\".hwe\"\n",
    "\t\n",
    "#### calculate MAF, make a list of snps with MAF > threshold in each pop\n",
    "for x in pops_to_keep:\n",
    "    print\"plink --bfile  fam_\" ,x, \"--maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist --autosome-num 50  --out fam_\" ,x, \".maf\"\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#plink2.bat\n",
    "#may 8, 2015\n",
    "\n",
    "plink --bfile fam_1 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_1.hwe\n",
    "plink --bfile fam_2 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_2.hwe\n",
    "plink --bfile fam_3 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_3.hwe\n",
    "plink --bfile fam_4 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_4.hwe\n",
    "plink --bfile fam_5 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_5.hwe\n",
    "plink --bfile fam_6 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_6.hwe\n",
    "plink --bfile fam_7 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_7.hwe\n",
    "plink --bfile fam_8 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_8.hwe\n",
    "plink --bfile fam_9 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_9.hwe\n",
    "plink --bfile fam_10 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_10.hwe\n",
    "plink --bfile fam_11 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_11.hwe\n",
    "plink --bfile fam_12 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_12.hwe\n",
    "plink --bfile fam_13 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_13.hwe\n",
    "plink --bfile fam_14 --hwe .05 midp --hardy midp --allow-extra-chr --allow-no-sex --write-snplist  --out fam_14.hwe\n",
    "plink --bfile fam_1 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_1.maf\n",
    "plink --bfile fam_2 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_2.maf\n",
    "plink --bfile fam_3 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_3.maf\n",
    "plink --bfile fam_4 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_4.maf\n",
    "plink --bfile fam_5 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_5.maf\n",
    "plink --bfile fam_6 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_6.maf\n",
    "plink --bfile fam_7 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_7.maf\n",
    "plink --bfile fam_8 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_8.maf\n",
    "plink --bfile fam_9 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_9.maf\n",
    "plink --bfile fam_10 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_10.maf\n",
    "plink --bfile fam_11 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_11.maf\n",
    "plink --bfile fam_12 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_12.maf\n",
    "plink --bfile fam_13 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_13.maf\n",
    "plink --bfile fam_14 --maf  0.05 --allow-extra-chr --allow-no-sex --write-snplist   --out fam_14.maf\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4\n",
    "### Apply filters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#step4.py\n",
    "#May 8, 2015\n",
    "\n",
    "import pandas as pd\n",
    "import collections\n",
    "\n",
    "pops_to_keep = ['1','2','3','4','5','6','7','8','9','10','11','12','13','14']\n",
    "\n",
    "passing_MAF = collections.defaultdict(int)\n",
    "\n",
    "for x in pops_to_keep:\n",
    "\twith open(\"fam_{}.maf.snplist\".format(x)) as INFILE:\n",
    "\t\tfor line in INFILE:\n",
    "\t\t\tpassing_MAF[line.strip()] +=1\n",
    "\n",
    "with open(\"passing_MAF.snplist\", 'w') as OUTFILE:\n",
    "\tfor locus in passing_MAF.keys():\n",
    "\t\tOUTFILE.write(locus + \"\\n\")\n",
    "\n",
    "passing_HWE = collections.defaultdict(int)\n",
    "\n",
    "for x in pops_to_keep:\n",
    "\twith open(\"fam_{}.hwe.snplist\".format(x)) as INFILE:\n",
    "\t\tfor line in INFILE:\n",
    "\t\t\tpassing_HWE[line.strip()] +=1\n",
    "\n",
    "with open(\"passing_HWE.snplist\", 'w') as OUTFILE:\n",
    "\tfor locus, pass_cnt in passing_HWE.items():\n",
    "\t\tif pass_cnt >= 5:\n",
    "\t\t\tOUTFILE.write(locus + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "####remove loci failing HWE or MAF filters locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile  miss_final --extract passing_MAF.snplist --allow-extra-chr --allow-no-sex --write-snplist --make-bed --out filter_MAF\n",
    "plink --bfile  filter_MAF --extract passing_HWE.snplist --allow-extra-chr --allow-no-sex --write-snplist  --make-bed --out filter_HWE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### global MAF filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile  filter_HWE --maf .05 --allow-extra-chr --allow-no-sex --write-snplist --freq  --make-bed --out filter_HWE_MAF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 5 \n",
    "### Retain a single SNP at each locus\n",
    "#### select based on maximum MAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import collections\n",
    "\n",
    "# Step 6 \n",
    "## Retain a single SNP at each locus\n",
    "### select based on maximum MAF\n",
    "\n",
    "maf_frq = pd.read_csv(\"filter_HWE_MAF.frq\", sep=\" \", skipinitialspace = True)\n",
    "\n",
    "maf_frq['catID']  =  [int(xx[0]) for xx in maf_frq['SNP'].str.split(\"_\", n = 1)]\n",
    "maf_frq['pos']  = [int(xx[1]) for xx in maf_frq['SNP'].str.split(\"_\", n = 1)]\n",
    "maf_frq.sort(['catID', 'MAF'], ascending = False, inplace = True)\n",
    "single_catID = maf_frq.drop_duplicates(subset = ['catID'])\n",
    "single_catID['SNP'].to_csv('single_catID.snplist', index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile  filter_HWE_MAF --extract single_catID.snplist --allow-extra-chr --allow-no-sex --write-snplist --make-bed --recode 12 A tabx --out complete"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Step 6\n",
    "###Summary Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get summary statistics on the final, filtered data set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile complete --family --missing --freq --het small-sample --ibc --fst --allow-extra-chr --allow-no-sex --write-snplist --out complete.1\n",
    "\n",
    "plink --bfile complete --family --freqx --allow-extra-chr --allow-no-sex --write-snplist --out complete.2\n",
    "\n",
    "plink --bfile complete --allow-extra-chr --allow-no-sex --write-snplist --recode A --out complete.3\n",
    "\n",
    "plink --bfile complete --family --missing --freq --het small-sample --ibc --fst --allow-extra-chr --allow-no-sex --write-snplist --out complete.4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##PCA in plink"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To make an individual PCA of the population data:\n",
    "\n",
    "The first step is to generate the compartison matrix using the output of the final filtering step. \n",
    "\n",
    "--make-rel is the primary interface to PLINK 1.9's realized relationship matrix and covariance matrix calculator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plink --bfile complete --make-rel --allow-extra-chr --allow-no-sex --write-snplist --out complete_rel\n",
    "\n",
    "plink --bfile complete  --pca --allow-extra-chr --allow-no-sex --write-snplist --out complete_pca"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I took the results into R to plot them with the package ggplot2\n",
    "\n",
    "The eigenvector file was merged with a little file that had a list of the population names, their population numbers and the lineage classification as a way to plot them by category. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "population_number\tpopulation_name\tlineage\n",
    "1\tPAMUR10\teven\n",
    "2\tPAMUR11\todd\n",
    "3\tPHAYLY09\todd\n",
    "4\tPHAYLY10\teven\n",
    "5\tPKOPE91T\todd\n",
    "6\tPKOPE96T\teven\n",
    "7\tPKUSHI06\teven\n",
    "8\tPKUSHI07\todd\n",
    "9\tPNOME91\todd\n",
    "10\tPNOME94\teven\n",
    "11\tPSNOH03\todd\n",
    "12\tPSNOH96\teven\n",
    "13\tPTAUY09\todd\n",
    "14\tPTAUY12\teven"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is the R code I used, it was written by Ryan."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#install.packages(\"ggplot2\", type=\"source\")\n",
    "\n",
    "library(ggplot2)\n",
    "\n",
    "eigenvec_table <- read.table(\"G:/Analysis/Pop_analysis/RenamedCatalog2/FilteringDips/complete_pca.eigenvec\")\n",
    "pop_key <- read.table(\"G:/Analysis/Pop_analysis/RenamedCatalog2/FilteringDips/pop_key.txt\", header = TRUE)\n",
    "\n",
    "tt = merge(eigenvec_table,pop_key, by.x = 'V1', by.y = 'population_number')\n",
    "\n",
    "ggplot(data = tt) + geom_point(aes(x = V3, y = V4,color = population_name), alpha = .5, size = 4) +\n",
    "   scale_color_discrete()\n",
    "ggsave()\n",
    "\n",
    "ggplot(data = tt) + geom_point(aes(x = V3, y = V4,color = lineage), alpha = .5, size = 4) +\n",
    "  scale_color_discrete()\n",
    "\n",
    "ggplot(data = tt) + geom_point(aes(x = V4, y = V7,color = population_name), alpha = .5, size = 4)\n",
    "\n",
    "ggplot(data = tt) + geom_point(aes(x = V3, y = V4,color = lineage), alpha = .5, size = 4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
