{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Summary Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Allele frequencies and allele counts \n",
    "\n",
    "This script that Wes wrote takes the output of genepop and a list of the population names and gives you the allele frequencies and the allele counts pere locus. \n",
    "\n",
    "perl genepop_al_freqs+al_counts.pl batch3_16681_oneperline_80.txt.INF batch3_16681_popnames_80.txt >16681_80_alfreq_alcounts.txt\n",
    "\n",
    "here are my pop names: \n",
    "\n",
    "    AMUR10\n",
    "    AMUR11\n",
    "    HAYLY09\n",
    "    HAYLY10\n",
    "    KOPPE91\n",
    "    KOPPE96\n",
    "    KUSHI06\n",
    "    KUSHI07\n",
    "    NOME91\n",
    "    NOME94\n",
    "    SNOH03\n",
    "    SNOH96\n",
    "    TAUY09\n",
    "    TAUY12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#genepop_al_freq+al_counts.pl\n",
    "\n",
    "#parsing genepop files to get allele frequency and allele counts (2N)\n",
    "#this script should be generalized and can probably be used for everything\n",
    "#usage:\n",
    "\n",
    "#perl genepop....pl genepop_al_freq_file.txt.inf popnames.txt > out.txt\n",
    "#argument 1: a .inf file from genepop\n",
    "#argument 2: a file with your popnames in order, one on each line, with no spaces, see example file\n",
    "\n",
    "print(\"Locus\\t\");\n",
    "\n",
    "open(POPS, \"<$ARGV[1]\") or die \"Error!!! reading $ARGV[1] for reading\";\n",
    "#makes a header with pop1 \\t pop2\\t pop1\\t pop2\\n \n",
    "@pop_names=();\n",
    "while(<POPS>)\n",
    "{\n",
    "  chomp($_);\n",
    "  $pop=\"$_\\t\";\n",
    "  push(@pop_names, $pop);\n",
    "}\n",
    "close POPS;\n",
    "\n",
    "print(\"@pop_names\");\n",
    "print(\"@pop_names\\n\");\n",
    "\n",
    "$num_pops=@pop_names;\n",
    "\n",
    "\n",
    "\n",
    "open(IN, \"<$ARGV[0]\") or die \"Error!!! reading $ARGV[0] for reading\"; #opens genepop file\n",
    "while(<IN>)\n",
    " {\n",
    "    chomp($_);\n",
    "    $line=$_;\n",
    "    if($line =~ m/^[:space:]*.Locus/) #looking for tables where lines start with Locus (towards end of file)\n",
    "    {\n",
    "      @title_line=split(/[:blank:]/ , $line); \n",
    "      chomp($title_line[1]);\n",
    "      print(\"$title_line[1]\\t\"); #this is just grabbing the SNP id\n",
    "      $_=<IN>; $_=<IN>; $_=<IN>; #skip 3 lines, this shouldn't have to be changed\n",
    "       \n",
    "      #if the data for this snp is not missing meaning it has ------ on the line after pop \\t alleles\\t genes\n",
    "      if($_ =~ m/-----/) #ex of missing data is snp 427 in WAK snp discov is basically Locus: \\n-----\\n No data \n",
    "      {\n",
    "      $_=<IN>; #skip a line\n",
    "      @al_freqs=();\n",
    "      @al_counts=();\n",
    "      #this may not work well for snps that are monomorphic in all pops but we don't want them anyway\n",
    "      for($i=1; $i<=$num_pops; $i=$i+1) #iterates through each pop for each snp snp grabbing freqs and counts\n",
    "      {\n",
    "        $freq_line=<IN>;\n",
    "        @split_freq_line=split(\" \" , $freq_line); #split by spaces \n",
    "        $each_freq=\"$split_freq_line[1]\\t\"; #grabbing the freq of al1\n",
    "        $each_count=\"$split_freq_line[3]\\t\"; #grabbing the counts\n",
    "        push(@al_freqs,$each_freq); \n",
    "        push(@al_counts,$each_count);\n",
    "        \n",
    "      }\n",
    "      \n",
    "      print(\"@al_freqs\");\n",
    "      print(\"@al_counts\\n\");\n",
    "           \n",
    "     }\n",
    "     else #if there is missing data just print snpid with nothing else\n",
    "      {\n",
    "        print(\"$title_line[1]\\n\");\n",
    "        redo;\n",
    "      }\n",
    "     }\n",
    "}\n",
    "close IN;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counts of Heterozygotes and missing genotypes from Genepop files. \n",
    "\n",
    "I ran the countHets.pl and the countMissing.pl programs on the genepop files. In one set of both files, I had to change the 'pop' to match the formatting in my genepop folder: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#countMissing.pl\n",
    "#garrett wrote this one\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "my$header=<>;\n",
    "chomp $header;\n",
    "my($col1,$col2,$samples)=split \"\\t\", $header, 3;\n",
    "my%missing;\n",
    "my@samples=split \"\\t\", $samples;\n",
    "while (my$line=<>){\n",
    "\tchomp $line;\n",
    "\tmy($tag,$genoNum,$rest)=split \"\\t\", $line, 3;\n",
    "\tmy@genos=split \"\\t\", $rest;\n",
    "\tforeach my $i (0..$#samples){\n",
    "\t\tif($genos[$i] eq \"-\"){\n",
    "\t\t\t$missing{$samples[$i]}++;\n",
    "\t\t}\n",
    "\t}\n",
    "}\n",
    "\n",
    "foreach my$key (keys %missing){\n",
    "\tprint \"$key\\t$missing{$key}\\n\";\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#countMissing.pl\n",
    "\n",
    "#!/usr/perl -w\n",
    "use strict;\n",
    "\n",
    "my$samples=0;\n",
    "print \"Sample\\tGenotyped Loci\\tMissing Loci\\tPercent Genotyped\\n\";\n",
    "while(my$line=<>){\n",
    "\tchomp $line;\n",
    "\tif($line eq \"pop\"){\n",
    "\t\t$samples=1;\n",
    "\t\tnext;\n",
    "\t}elsif($samples==1){\n",
    "\t\tmy@columns=split \"\\t\", $line;\n",
    "\t\tmy$sample=$columns[0];\n",
    "\t\tmy$missing=0;\n",
    "\t\tmy$columns=scalar@columns;\n",
    "\t\tmy$lociNum=$columns-1;\n",
    "\t\tforeach my$i (1..$#columns){\n",
    "\t\t\tif($columns[$i] eq \"0000\"){\n",
    "\t\t\t\t$missing++;\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t\tmy$genotyped=$lociNum-$missing;\n",
    "\t\tmy$percGeno=100-($missing/$lociNum*100);\n",
    "\t\tprint \"$sample:\\t$genotyped\\t$missing\\t$percGeno\\n\";\n",
    "\t}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#countHets.pl \n",
    "#garrett wrote this one\n",
    "\n",
    "#!/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "my$header=<>;\n",
    "chomp $header;\n",
    "my($col1,$col2,$samples)=split \"\\t\", $header, 3;\n",
    "my%missing;\n",
    "my@samples=split \"\\t\", $samples;\n",
    "while (my$line=<>){\n",
    "\tchomp $line;\n",
    "\tmy($tag,$genoNum,$rest)=split \"\\t\", $line, 3;\n",
    "\tmy@genos=split \"\\t\", $rest;\n",
    "\tforeach my $i (0..$#samples){\n",
    "\t\tif($genos[$i] =~ /\\//){\n",
    "\t\t\t$missing{$samples[$i]}++;\n",
    "\t\t}\n",
    "\t}\n",
    "}\n",
    "\n",
    "foreach my$key (keys %missing){\n",
    "\tprint \"$key\\t$missing{$key}\\n\";\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#countHets.pl \n",
    "\n",
    "#!/usr/perl -w\n",
    "use strict;\n",
    "\n",
    "my$samples=0;\n",
    "print \"Sample\\tGenotyped Loci\\tHeterozygous Loci\\tPercent Heterozygous\\n\";\n",
    "while(my$line=<>){\n",
    "\tchomp $line;\n",
    "\tif($line eq \"pop\"){\n",
    "\t\t$samples=1;\n",
    "\t\tnext;\n",
    "\t}elsif($samples==1){\n",
    "\t\tmy@columns=split \"\\t\", $line;\n",
    "\t\tmy$sample=$columns[0];\n",
    "\t\tmy$missing=0;\n",
    "\t\tmy$columns=scalar@columns;\n",
    "\t\tmy$lociNum=$columns-1;\n",
    "\t\tforeach my$i (1..$#columns){\n",
    "\t\t\tif($columns[$i] eq \"0000\"){\n",
    "\t\t\t\t$missing++;\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t\tmy$het=0;\n",
    "\t\tforeach my$i (1..$#columns){\n",
    "\t\t\tmy$allele1=substr $columns[$i], 0, 2;\n",
    "\t\t\tmy$allele2=substr $columns[$i], 2, 2;\n",
    "\t\t\tif($allele1 ne $allele2){\n",
    "\t\t\t\t$het++;\n",
    "\t\t\t}\n",
    "\t\t\t#print \"$allele1\\t$allele2\\n\";\n",
    "\t\t}\n",
    "\t\tmy$genotyped=$lociNum-$missing;\n",
    "\t\tmy$percGeno=100-($missing/$lociNum*100);\n",
    "\t\tmy$percHet=($het/$lociNum*100);\n",
    "\t\tmy$percHetGeno=($het/$genotyped*100);\n",
    "\t\t#print \"$sample:\\t$genotyped\\t$missing\\t$percGeno\\n\";\n",
    "\t\tprint \"$sample:\\t$genotyped\\t$het\\t$percHetGeno\\n\";\n",
    "\t}\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Garrett wrote some perl script that takes a genepop input file and gives you among other things the observed heterozygosity per locus by population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#/usr/bin/perl -w\n",
    "\n",
    "#To run it follow this example: perl countHetsMissing_genepop_byPop.pl inputfile >outputfile\n",
    "use strict;\n",
    "\n",
    "my$header=<>;\n",
    "\n",
    "my@loci;\n",
    "my%missingGenos;\n",
    "my%hetGenos;\n",
    "my%totalGenos;\n",
    "my$locusSection=1;\n",
    "my$pop=0;\n",
    "print \"Pop\\tID\\tTotalGenos\\tMissingGenos\\tHeterozygotes\\tMissingPerc\\tHetPerc\\n\";\n",
    "while(my$line=<>){\n",
    "\tchomp $line;\n",
    "\tif($line=~/pop/i){\n",
    "\t\t$locusSection=0;\n",
    "\t\tif($pop==0){\n",
    "\t\t\t$pop++;\n",
    "\t\t\tnext;\n",
    "\t\t}else{\n",
    "\t\t\tforeach my$key (sort keys %totalGenos){\n",
    "\t\t\t\tmy$missingPerc;\n",
    "\t\t\t\tmy$hetPerc;\n",
    "\t\t\t\tif(not exists $missingGenos{$key}){\n",
    "\t\t\t\t\t$missingGenos{$key}=0;\n",
    "\t\t\t\t\t$missingPerc=0\n",
    "\t\t\t\t}else{\n",
    "\t\t\t\t\t$missingPerc=$missingGenos{$key}/$totalGenos{$key};\n",
    "\t\t\t\t}\n",
    "\t\t\t\tif(not exists $hetGenos{$key}){\n",
    "\t\t\t\t\t$hetGenos{$key}=0;\n",
    "\t\t\t\t\t$hetPerc=0;\n",
    "\t\t\t\t}else{\n",
    "\t\t\t\t\t$hetPerc=$hetGenos{$key}/$totalGenos{$key};\n",
    "\t\t\t\t}\n",
    "\t\t\t\tprint \"$pop\\t$key\\t$totalGenos{$key}\\t$missingGenos{$key}\\t$hetGenos{$key}\\t$missingPerc\\t$hetPerc\\n\";\n",
    "\t\t\t}\n",
    "\t\t\t%missingGenos=();\n",
    "\t\t\t%hetGenos=();\n",
    "\t\t\t%totalGenos=();\n",
    "\t\t\t$pop++;\n",
    "\t\t\tnext;\n",
    "\t\t}\n",
    "\t}\n",
    "\tif($locusSection==1){\n",
    "\t\tpush@loci,$line;\n",
    "\t}else{\n",
    "\t\tmy@columns=split \"\\t\", $line;\n",
    "\t\tshift@columns;\n",
    "\t\tforeach my$i (0..$#columns){\n",
    "\t\t\tmy$allele1=substr $columns[$i], 0, 2;\n",
    "\t\t\tmy$allele2=substr $columns[$i], 2, 2;\n",
    "\t\t\tif($allele1 eq \"00\"){\n",
    "\t\t\t\t$missingGenos{$loci[$i]}++;\n",
    "\t\t\t}elsif($allele1 ne $allele2){\n",
    "\t\t\t\t$hetGenos{$loci[$i]}++;\n",
    "\t\t\t\t$totalGenos{$loci[$i]}++;\n",
    "\t\t\t}else{\n",
    "\t\t\t\t$totalGenos{$loci[$i]}++;\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "}\n",
    "\n",
    "foreach my$key (sort keys %totalGenos){\n",
    "\tmy$missingPerc;\n",
    "\tmy$hetPerc;\n",
    "\tif(not exists $missingGenos{$key}){\n",
    "\t\t$missingGenos{$key}=0;\n",
    "\t\t$missingPerc=0\n",
    "\t}else{\n",
    "\t\t$missingPerc=$missingGenos{$key}/$totalGenos{$key};\n",
    "\t}\n",
    "\tif(not exists $hetGenos{$key}){\n",
    "\t\t$hetGenos{$key}=0;\n",
    "\t\t$hetPerc=0;\n",
    "\t}else{\n",
    "\t\t$hetPerc=$hetGenos{$key}/$totalGenos{$key};\n",
    "\t}\n",
    "\tprint \"$pop\\t$key\\t$totalGenos{$key}\\t$missingGenos{$key}\\t$hetGenos{$key}\\t$missingPerc\\t$hetPerc\\n\";\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I took the results of those into excel and formatted them to line up. Now I have allele frequencies, counts, missing loci, percent genotyped loci, and percent heterozygous loci per population. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#*****Summary stats also in the ipython notebook _Filtered_in_PLINK*****\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##The sample betweeen Snohomish and Koppen that is a hybrid? \n",
    "\n",
    "One sample in the Snohomish 96 population looks like its a hybrid with the Koppen. If I plot the North America samples alone or the even-year lineage alone, it plots right between the Koppen population and the Snohomish 96.\n",
    "\n",
    "It's sample PSNOH96_0028.\n",
    "\n",
    "I looked at the summary stats for the sample, but there was nothing out of the ordinary. It was average when you look at heterozygosity, and the percent genotyped was 98% so that's not an issue either. I will not remove the sample; it appears to be legitimate, and could be a true hybrid. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Genepop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#get_HWresults_from_genepop.pl\n",
    "\n",
    "#/usr/bin/perl -w\n",
    "use strict;\n",
    "\n",
    "my$firstLocus=0;\n",
    "my$getPval=0;\n",
    "while(my$line=<>){\n",
    "\tchomp $line;\n",
    "\tif($line=~/^-/){\n",
    "\t\tnext;\n",
    "\t}\n",
    "\tif($line=~/Locus/){\n",
    "\t\t$firstLocus++;\n",
    "\t\tmy($discard,$locus)=split \" \", $line, 2;\n",
    "\t\t$locus=~s/\"//g;\n",
    "\t\tprint \"\\n$locus\";\n",
    "\t}\n",
    "\tif($getPval==1){\n",
    "\t\tmy($pop,$pval,$rest)=split \" \", $line, 3;\n",
    "\t\tprint \"\\t$pval\";\n",
    "\t\t#print \"$pval\";\n",
    "\t}\n",
    "\tif($line=~/POP/){\n",
    "\t\t$getPval=1;\n",
    "\t}\n",
    "\tif($line=~/^ *$/){\n",
    "\t\t$getPval=0;\n",
    "\t\t#print \"\\n\";\n",
    "\t}\n",
    "\t#print \"|\";\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adegenet\n",
    "\n",
    "For running some basic summary statistics, doing AMOVA and a dapc\n",
    "    #Adegenet.R      7/22/2015 and 9/2/2015 \n",
    "    \n",
    "### Input file: \n",
    "    \n",
    "Genepop file of the filtered 16681 at the individuals that had a genotype rate of at least 80%. The filtered genepop file did not open in R, so I used the 16681 genepop file that didn't have the individuals filtered out, and when that one loaded in R, I opened it up and took out these individuals that didn't pass the threshold: \n",
    "    \n",
    "- Hayly09_32\n",
    "- Hayly10_14\n",
    "- Nome94_12\n",
    "- Nome94_09\n",
    "- Tauy09_32\n",
    "- Tauy09_21\n",
    "- Tauy09_24\n",
    "- Tauy09_09\n",
    "- Tauy12_01\n",
    "- Tauy12_06\n",
    "- Tauy12_14\n",
    "- Tauy12_33\n",
    "- Tauy12_38\n",
    "\n",
    "    I saved it as 16681_80_edit.gen and imported it with that name into the program. \n",
    "    \n",
    "Here is a summary of the data from the program: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "> pops_16681_edit\n",
    "/// GENIND OBJECT /////////\n",
    "\n",
    " // 383 individuals; 16,681 loci; 49,914 alleles; size: 82.8 Mb\n",
    "\n",
    " // Basic content\n",
    "   @tab:  383 x 49914 matrix of allele counts\n",
    "   @loc.n.all: number of alleles per locus (range: 2-3)\n",
    "   @loc.fac: locus factor for the 49914 columns of @tab\n",
    "   @all.names: list of allele names for each locus\n",
    "   @ploidy: ploidy of each individual  (range: 2-2)\n",
    "   @type:  codom\n",
    "   @call: read.genepop(file = \"G:/Analysis/Pop_analysis/Populations_b3_may/Adegenet/16681_80_edit.gen\")\n",
    "\n",
    " // Optional content\n",
    "   @pop: population of each individual (group size range: 18-32)          \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Run Adegenet on the population data\n",
    "###   for summary stats, AMOVA, dapc\n",
    "### Carolyn Tarpey | September 2015 \n",
    "### ---------------------------------------\n",
    "\n",
    "\n",
    "install.packages(\"adegenet\", dep=TRUE)\n",
    "install.packages(\"poppr\", dep =TRUE)\n",
    "\n",
    "\n",
    "library(adegenet)\n",
    "library(ape)\n",
    "library(pegas)\n",
    "library(seqinr)\n",
    "library(ggplot2)\n",
    "library(poppr)\n",
    "library(hierfstat)\n",
    "\n",
    "\n",
    "##this one works, but the data has not been filtered for individuals\n",
    "pops_16681 <- read.genepop(\"G:/Analysis/Pop_analysis/Populations_b3_may/Adegenet/16681_nopop_no80.gen\")\n",
    "\n",
    "#This one works too and is the final data set\n",
    "pops_16681_edit <- read.genepop(\"G:/Analysis/Pop_analysis/Populations_b3_may/Adegenet/16681_80_edit.gen\")\n",
    "\n",
    "pops_16681_edit@pop\n",
    "\n",
    "names(pops_16681_edit)\n",
    "\n",
    "toto <- summary(pops_16681_edit)\n",
    "\n",
    "edit(pops_16681_edit)\n",
    "names(toto)\n",
    "\n",
    "## Q: Is the Mean observed H significantly lower than the mean expected H? \n",
    "\n",
    "bartlett.test(list(toto$Hexp, toto$Hobs))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "> bartlett.test(list(toto$Hexp, toto$Hobs))\n",
    "\n",
    "\tBartlett test of homogeneity of variances\n",
    "\n",
    "data:  list(toto$Hexp, toto$Hobs)\n",
    "Bartlett's K-squared = 116.93, df = 1, p-value < 2.2e-16\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t.test(toto$Hexp,toto$Hobs, pair = T, var.equal = TRUE, alter = \"greater\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "> t.test(toto$Hexp,toto$Hobs, pair = T, var.equal = TRUE, alter = \"greater\")\n",
    "\n",
    "\tPaired t-test\n",
    "\n",
    "data:  toto$Hexp and toto$Hobs\n",
    "t = 163.84, df = 16680, p-value < 2.2e-16\n",
    "alternative hypothesis: true difference in means is greater than 0\n",
    "95 percent confidence interval:\n",
    " 0.04644439        Inf\n",
    "sample estimates:\n",
    "mean of the differences \n",
    "             0.04691542 \n",
    "\t\t\t "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#This is taking forever to run \n",
    "pairwise.fst(pops_16681_edit, pop= NULL, res.type = c(\"dist\", \"matrix\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "> pairwise.fst(pops_16681_edit, pop= NULL, res.type = c(\"dist\", \"matrix\"))\n",
    "            1          2          3          4          5          6          7          8          9         10         11         12         13\n",
    "2  0.03713108                                                                                                                                    \n",
    "3  0.03578918 0.01066476                                                                                                                         \n",
    "4  0.01221767 0.03705546 0.03560100                                                                                                              \n",
    "5  0.05044387 0.02753836 0.02758697 0.04996322                                                                                                   \n",
    "6  0.02373888 0.04728466 0.04579836 0.02243125 0.05672948                                                                                        \n",
    "7  0.01378259 0.04147544 0.04081558 0.01926097 0.05490709 0.03093015                                                                             \n",
    "8  0.04442862 0.01364543 0.01718015 0.04505148 0.03260416 0.05492264 0.04877038                                                                  \n",
    "9  0.03663895 0.01211269 0.01108887 0.03654334 0.02965242 0.04874063 0.04172586 0.01863162                                                       \n",
    "10 0.01681382 0.03823536 0.03730709 0.01340815 0.05482867 0.02718595 0.02274181 0.04622084 0.04082480                                            \n",
    "11 0.06849168 0.04433646 0.04540019 0.06731200 0.03558518 0.07585488 0.07109011 0.04972008 0.04760146 0.07065036                                 \n",
    "12 0.03469519 0.05091818 0.05060942 0.03275280 0.05994755 0.02813025 0.03828925 0.05899044 0.05192792 0.03442890 0.07182033                      \n",
    "13 0.03591335 0.01107856 0.01039770 0.03568867 0.02850251 0.04682780 0.04060512 0.01754976 0.01205706 0.03832820 0.04595166 0.05020727           \n",
    "14 0.01208570 0.03693895 0.03532381 0.01046837 0.05122340 0.02395459 0.01880507 0.04483438 0.03694565 0.01481165 0.06871101 0.03375420 0.03551611\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "#im not really sure what this is, a discriminant analysis of principal components \n",
    "dapc.All<- dapc (pops_16681_edit, pops_16681_edit$pop)\n",
    "??dapc\n",
    "vignette(\"adegenet-dapc\")\n",
    "\n",
    "\n",
    "#####################################\n",
    "###The rest of this code is derived from here and google searches on the internet\n",
    "##http://grunwaldlab.github.io/Population_Genetics_in_R/DAPC.html\n",
    "\n",
    "\n",
    "scatter(dapc.All, cell = 0, posi.da = \"topleft\", \n",
    "        scree.pca= TRUE, posi.pca= \"bottomleft\", ratio.da = .2, cex.lab = .5,\n",
    "        ratio.pca= .2, pch = 10, cstar = 1,  axesell = TRUE, mstree = TRUE, lwd = 2, lty = 2)\n",
    "\n",
    "names(dapc.All)\n",
    "\n",
    "set.seed(14)\n",
    "\n",
    "# change graphical parameter to subsequently overlay the labels withoutdrawing a new plot\n",
    "par(new=TRUE)\n",
    "# make a data frame of the dapc coordinates used in scatter\n",
    "df <- data.frame(x = dapc.All$ind.coord[,1], y = dapc.All$ind.coord[,2])\n",
    "# identify/ create a vector of names for the individuals in your plot\n",
    "noms <- paste(\"ind\", c(1:400), sep=\".\")\n",
    "# use the text function to add labels to the positions given by the coordinates you used in plot\n",
    "s.label(dfxy = df, xax=1, yax=2, label=noms,\n",
    "        clabel=0.7, # change the size of the labels\n",
    "        boxes=TRUE, # if points are spaced wide enough, can use TRUE to add boxes around the labels\n",
    "        grid=FALSE, addaxes=FALSE) # do not draw lines or axes in addition to the labels\n",
    "\n",
    "\n",
    "\n",
    "names(pops_16681_edit)\n",
    "pops_16681_edit$loc.fac\n",
    "\n",
    "\n",
    "contrib <- loadingplot(dapc.All$var.contr, axis= 2, thresh= .001,  lab.jitter= 1)\n",
    "\n",
    "temp <- seploc(pops_16681_edit)\n",
    "\n",
    "names(temp) #lists all the loci names\n",
    "\n",
    "dapc.All$var.contr\n",
    "\n",
    "dapc.df = print.dapc(dapc.All)\n",
    "\n",
    "write.table(dapc.All, file= \"G:/Analysis/Pop_analysis/Populations_b3_may/Adegenet/16681_R_dapc.txt\", sep = '\\t', eol = \"\\n\", row.names = TRUE, col.names= TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "September 2 \n",
    "\n",
    "I started an DAPC in ADEGENET; it finished running the pairwise FST and I asked it to start the DAPC, and it asked me to choose the number of PCs to retain and gave me a plot of the cumulative variance explained by each PC.\n",
    "\n",
    "Ryan's said that DAPCs don't use the genotypes of the idividuals, they also look at the groupings that you gave it, so it looks for which axises can explain the variation in the groups that you have assignd, and for that reason it has some circular logic in the priors that you have given it. He says to think about why you would want to do a test like this, and what you expect to get out of it. Maybe you could use it to find the snps that have the heavy loadings in groupings, and maybe it does a better job of differentiating the groups that have some difficulty being differntiatied, in my case this could be the Asia and Nome groups? \n",
    "\n",
    "But most of all, the number of pcs that you tell it can have a huge impact on the outcome of the test, so use some sort of statistical method to figure out a significance for the pcs you choose. He said that there's something called a scree plot that has the percent variance explained per factor, but they aren't cumulative, so you can see the bend in the graph. He also said that if it doesnt give a scree plot, that you're looking for an asymptote in the other plot. But I dont really  have much of one? So ive got to do some reading in the manual about what it exactly is looking for. \n",
    "\n",
    "i plotted the loadings and that was pretty cool, there were a couple of loci that really stood out, though whether they were actually significant, Im not sure how to judge. \n",
    "\n",
    "I plotted the scatter plot and messed with some of the formatting and also looked at the loadings, but couldnt figure out which of the loci were actually the ones above my arbitrary threshold.\n",
    "    \n",
    "    \n",
    "         $$ September 3 2015\n",
    "\n",
    "\n",
    "talked to garrett today, showed him the work id done in Adegenet, and he said that the names in the loadings plot are actually the rad tag names, with some kind of suffix, and he doesnt know what the suffix is, maybe the axis it is important on? \n",
    "\n",
    "anyway, this threshold i made is arbitrary, but i could export the loadings and put them into a distribution, and from there do a statistical test, like look at what kindof distribution they make then see where they fall out on the distribution and see if that is statistically important. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Arlequin \n",
    "\n",
    "July 15\n",
    "\n",
    "Lisa's paper says that they did their amova in arlequin, they did two, one grouping populations within lineages and one grouping within sites to determine the greatest source of overall genetic variation. \n",
    "\n",
    "I used PGDSpider to translate the genepop file batch_3_16681_oneline_80 to 16681_80.arp\n",
    "\n",
    "I changed the names in the file to the Sili River names and saved it.\n",
    "\n",
    "Using Arlequin, I: \n",
    "\n",
    "\tDetecting loci under selection \n",
    "\t\tuse hierarchical island model\n",
    "\t\t\t20,000 simulations\n",
    "\t\t\t100 demes\n",
    "\t\t\t14 populations\n",
    "\t\t\tmin exp het 0 \n",
    "\t\t\tmax epe het 1\n",
    "\t\t\tdistance method for AMOVA computations: number of different allele\n",
    "\tpopulation comparisons\n",
    "\tpopulation differentiation\n",
    "\tlinkage disequilibrium\n",
    "\t\tHW\n",
    "\t\tpairwise linkage\n",
    "\tMolecular diversity indices\n",
    "    \n",
    "july 20\n",
    "\n",
    "I looked at the Arlequin, it didnt run because of some error to do with the imput files. Garrett gave me one of his to look at to see if mine looks normal. Wes also said to count up the number of samples on the line with the population name or something. \n",
    "\n",
    "\n",
    "I opened and changed the names of the populations to the actual names of the populations and saved the file with the suffix -named and then loaded it into Arlequin. i went to the structure editor and changed the structure of the populations, so that the even were a 2 and the odd were a 1. i havent really got what the amova is testing but I do know that you have to give it some structure before you can make the option to do an AMOVA even possible in the program. I hope that this distinction of the odd and even lineages is the best one, and that the one and twos dont mean some other kind of hierarchy that i dont realize. \n",
    "\n",
    "\n",
    "I started the AMOVA. and I didnt give it any other specifications for other things to do. when the popup asked what i wanted to do with the unphased data, i just left whatever the default was, the lower of the options, it was the one that i think basically was like ignore that they are unphased and if there is something that needs phased data, just dont do it. \n",
    "\n",
    "After reading Lisa's paper I changed the settings to be the locus by locus AMOVA, whereas before I had the standard amova and then did include individual level and changed the number of permutations to 10,000. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "================================================================\n",
    "  AMOVA ANALYSIS\n",
    "=================================================================\n",
    "--------------------------\n",
    "Genetic structure to test :\n",
    "---------------------------\n",
    "No. of Groups = 2\n",
    "\n",
    "\t[[Structure]]\n",
    "\t\tStructureName = \"New Edited Structure\"\n",
    "\t\tNbGroups = 2\n",
    "\t#Group1\n",
    "\t\tGroup={\n",
    "\t\t\"AMUR10\"\n",
    "\t\t\"HAYLY10\"\n",
    "\t\t\"KOPPE96\"\n",
    "\t\t\"KUSHI06\"\n",
    "\t\t\"NOME94\"\n",
    "\t\t\"SNOH96\"\n",
    "\t\t\"TAUY12\"\n",
    "\t\t}\n",
    "\n",
    "\t#Group2\n",
    "\t\tGroup={\n",
    "\t\t\"AMUR11\"\n",
    "\t\t\"HAYLY09\"\n",
    "\t\t\"KOPPE91\"\n",
    "\t\t\"KUSHI07\"\n",
    "\t\t\"NOME91\"\n",
    "\t\t\"SNOH03\"\n",
    "\t\t\"TAUY09\"\n",
    "\t\t}\n",
    "---------------------------\n",
    "Distance method: Pairwise differences\n",
    "--------------------------\n",
    "AMOVA design and results :\n",
    "--------------------------\n",
    "----------------------------------------------------------------------\n",
    " Source of                  Sum of      Variance         Percentage\n",
    " variation      d.f.        squares     components       of variation\n",
    "----------------------------------------------------------------------\n",
    " Among\n",
    " groups           1      33395.308       78.21504 Va             5.72\n",
    "\n",
    " Among\n",
    " populations\n",
    " within\n",
    " groups          12      40660.388       39.23909 Vb             2.87\n",
    "\n",
    " Within\n",
    " populations    752     940106.661     1250.14184 Vc            91.41\n",
    "----------------------------------------------------------------------\n",
    " Total          765    1014162.356     1367.59596\n",
    "----------------------------------------------------------------------\n",
    " Fixation Indices\n",
    "      FST :      0.08588\n",
    "      FSC :      0.03043\n",
    "      FCT :      0.05719\n",
    "----------------------------------------------------------------------\n",
    "\n",
    "Significance tests (1023 permutations)\n",
    "------------------\n",
    "\n",
    "Vc and FST : P(rand. value &lt; obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00000\n",
    "                                  P-value =  0.00000+-0.00000\n",
    "\n",
    "Vb and FSC : P(rand. value > obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00000\n",
    "                                  P-value =  0.00000+-0.00000\n",
    "\t\t\t\t\t\t\t\t  \n",
    "Va and FCT : P(rand. value > obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00098\n",
    "                                  P-value =  0.00098+-0.0009\n",
    "\t\t\t\t\t\t\t\t  \n",
    "////////////////////////////////////////////////////////////////////\n",
    "END OF RUN NUMBER 1 (21/07/15 at 15:15:53))\n",
    "Total computing time for this run : 0h 8m 48s 688 ms\n",
    "////////////////////////////////////////////////////////////////////\n",
    "\n",
    "Project information:\n",
    "--------------------\n",
    "\tNbSamples      =  14\n",
    "\tDataType       =  STANDARD\n",
    "\tGenotypicData  =  1\n",
    "\tGameticPhase   =  0\n",
    "\tRecessiveData  =  0\n",
    "\n",
    "==============================\n",
    "Settings used for Calculations\n",
    "==============================\n",
    " General settings:\n",
    " -----------------\n",
    "        Deletion Weight                = 1\n",
    "        Transition Weight Weight       = 1\n",
    "        Tranversion Weight Weight      = 1\n",
    "        Epsilon Value                  = 1e-07\n",
    "        Significant digits for output  = 5\n",
    "        Infer haplotype definition from distance matrix\n",
    "        Allowed level of missing data = 0.05\n",
    "\n",
    " Active Tasks:\n",
    " -------------\n",
    "    Analysis of Molecular Variance:\n",
    "    -------------------------------\n",
    "        No. of Permutations = 10000\n",
    "\n",
    "       Distance matrix:\n",
    "            Compute distance matrix\n",
    "            Molecular distance : Pairwise differences\n",
    "            Gamma a value      = 0\n",
    "\n",
    " #[ERROR # 15] : haplotype list is empty\n",
    " #[ERROR # 16] : haplotype list is empty\n",
    " #[ERROR # 17] : haplotype list is empty\n",
    " #[ERROR # 18] : haplotype list is empty\n",
    " #[ERROR # 19] : haplotype list is empty\n",
    " #[ERROR # 20] : haplotype list is empty\n",
    " #[ERROR # 21] : haplotype list is empty\n",
    " #[ERROR # 22] : haplotype list is empty\n",
    " #[ERROR # 23] : haplotype list is empty\n",
    " #[ERROR # 24] : haplotype list is empty\n",
    " #[ERROR # 25] : haplotype list is empty\n",
    " #[ERROR # 26] : haplotype list is empty\n",
    " #[ERROR # 27] : haplotype list is empty\n",
    " #[ERROR # 28] : haplotype list is empty\n",
    "\n",
    "===============================================================================\n",
    "== GENETIC STRUCTURE ANALYSIS\n",
    "===============================================================================\n",
    "\n",
    "********************\n",
    "Locus by locus AMOVA:\n",
    "********************\n",
    "Project information:\n",
    "--------------------\n",
    "\tNbSamples      =  14\n",
    "\tDataType       =  STANDARD\n",
    "\tGenotypicData  =  1\n",
    "\tGameticPhase   =  0\n",
    "\tRecessiveData  =  0\n",
    "\n",
    "==============================\n",
    "Settings used for Calculations\n",
    "==============================\n",
    "\n",
    " General settings:\n",
    " -----------------\n",
    "        Deletion Weight                = 1\n",
    "        Transition Weight Weight       = 1\n",
    "        Tranversion Weight Weight      = 1\n",
    "        Epsilon Value                  = 1e-07\n",
    "        Significant digits for output  = 5\n",
    "        Infer haplotype definition from distance matrix\n",
    "        Alllowed level of missing data = 0.05\n",
    "\n",
    " Active Tasks:\n",
    " -------------\n",
    "\n",
    "    Analysis of Molecular Variance:\n",
    "    -------------------------------\n",
    "        No. of Permutations = 10000\n",
    "\n",
    "       Distance matrix:\n",
    "            Compute distance matrix\n",
    "            Molecular distance : Pairwise differences\n",
    "            Gamma a value      = 0\n",
    "\n",
    " #[ERROR # 1] : haplotype list is empty\n",
    " #[ERROR # 2] : haplotype list is empty\n",
    " #[ERROR # 3] : haplotype list is empty\n",
    " #[ERROR # 4] : haplotype list is empty\n",
    " #[ERROR # 5] : haplotype list is empty\n",
    " #[ERROR # 6] : haplotype list is empty\n",
    " #[ERROR # 7] : haplotype list is empty\n",
    " #[ERROR # 8] : haplotype list is empty\n",
    " #[ERROR # 9] : haplotype list is empty\n",
    " #[ERROR # 10] : haplotype list is empty\n",
    " #[ERROR # 11] : haplotype list is empty\n",
    " #[ERROR # 12] : haplotype list is empty\n",
    " #[ERROR # 13] : haplotype list is empty\n",
    " #[ERROR # 14] : haplotype list is empty\n",
    "</data>\n",
    "\n",
    " <A NAME=\"21_07_15at17_12_57_pop_gen_struct\"></A> \n",
    "<data>\n",
    "\n",
    "===============================================================================\n",
    "== GENETIC STRUCTURE ANALYSIS\n",
    "===============================================================================\n",
    "\n",
    "\n",
    "********************\n",
    "Locus by locus AMOVA:\n",
    "********************\n",
    "\n",
    "Distance method for locus-by-locus analysis: Pairwise differences\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "July 23: \n",
    "    I started Arlequin again, this time with some basics, like:\n",
    " the molecular diversity indices, (standard diversity indices, output sample allele frequencies for all loci) \n",
    " exact test of population differentiation\n",
    " pairwise FST\n",
    "\tcompute pairwise differences (pi) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "july 26\n",
    "\n",
    "\n",
    "The outlier test in Arlequin; Garrett said to just run it and see what happens, then tweak the issues from there. Also that after you run it, to press the Rcmd button to get the plots in R, you may have to navigate to where you have R installed on the computer if it prompts you or you can go to the arlequin configurations and change it; the one you want is the Rcmd.exe file. \n",
    "\n",
    "have more groups than you actually do, so set it to something like 50 and then the rest of it you just run with the defaults. \n",
    "\n",
    "i started an AMOVA with North America and Asia as the groups and Nome in with North America its the one that is 24/07/15 at 12:13:15 \n",
    "\n",
    "I started another AMOVA with the Nome in with Asia at 12:57. \n",
    "\n",
    "I restructured to be odd and even, and started an AMOVA and an outlier test at 2:35. I kept all the defaults, but changed the number of groups to simulate to 50, because that is what it recommended. \n",
    "\n",
    "I talked to Ryan and garrett both about how to go about the test in arlequin that was supposed to replace the pairwise tests in bayescan, and both agreed that there's no way to get around it, the only way I have to answer that particular question is with pairwise tests. \n",
    "\n",
    "Ryan had this recommendation: \n",
    "\n",
    "1. arlequin run of odd vs even, using the hierarchical structure. \n",
    "2. pairwise bayescan for each river and compare them; in general, bayescan is better, so if there is any reason that you dont need the hierarchy, then go with Bayescan\n",
    "3.  bayescan run with all the populations as one thing. so it looks for the neutral ones\n",
    "\ttake the neutral loci and make a tree out of them\n",
    "4. bayescan with in each lineage, like and even and an odd. with the north america and the asia mixed together. \n",
    "5. bayenv2 scan and look for the overlaps with the bayescan \n",
    "\n",
    "ryan cautions that the outlier methods are not meant for these long time scale separations and large geographical regions. \n",
    "he says to think about whether the question is going to be about coonvergence or about the divergence. it will make the comparisons maybe more obvious. also think about how you dont just want to repeat mortens paper, you want to have some kindof spin or difference. \n",
    "\n",
    "july 27\n",
    "\n",
    "The other thing that ran over the weekend was the arlequin outlier tests. Garrett helped me get the graphs in R, they needed some functions installed from the Alrequin website to be replaced. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How should i do my grouping for the amova? like after i do the odd vs even, how do i do a hierarchical amova that takes into acount the continetal split (the answer is you probably cant) and how do i know whether to put nome in with asia or with north america? you can just look at your PCA and see what it groups with, or you can take a look at the individual assignment or mixture analysis or reporting groups. but that takes GISim and its not in the scope of this project so it might be overkill. the way that works though is to take a group and assign the individuals back to one or the other. and in this case you could do a continental split, where one run had nome with asia and another run had nome with north america and then you can take a look to see in which case the nome individuals were assigned the easiest. in gisim, you simulate new fish based oon the fish you have so you can run tons of different tests. if you only had one set of fish you could run the test once and you wouldnt know how good it was at assigning the fish back more than once. but if you simulate a ton more fish and have them tested lots, you get replication and have a much better idea of what is going on. \n",
    "\n",
    "so do an amova with nome in asia (the mst likely because of what the PCA says) and then do another PCA of the fish with North America and see if anything is different. if there isnt any difference, group the nome with asia because of the PCA. \n",
    "\n",
    "what you want to report with the AMOVA is the FCt overall, that tells you how different are these groups? \n",
    "\n",
    "the plots that look like this are telling you what your confidence intervals are: \n",
    "\n",
    "\n",
    "Average F-Statistics over all loci\n",
    " Fixation Indices\n",
    "      FST :      0.08449\n",
    "      FSC :      0.03031\n",
    "      FCT :      0.05587\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "FCT bootstrap percentile values (over 20000 bootstraps)\n",
    "-------------------------------------------------------\n",
    "\n",
    " 0.05 %    : 0.05283\n",
    " 0.1 %     : 0.05297\n",
    " 0.25 %    : 0.05316\n",
    " 0.5 %     : 0.05335\n",
    " 1 %       : 0.05360\n",
    " 2.5 %     : 0.05396\n",
    " 5 %       : 0.05427\n",
    " 10 %      : 0.05461\n",
    " 50 %      : 0.05585\n",
    " 90 %      : 0.05711\n",
    " 95 %      : 0.05747\n",
    " 97.5 %    : 0.05782\n",
    " 99 %      : 0.05823\n",
    " 99.5 %    : 0.05849\n",
    " 99.75 %   : 0.05875\n",
    " 99.9 %    : 0.05904\n",
    " 99.95 %   : 0.05928\n",
    "\n",
    "\n",
    "why are the amounts for the differences so small, and why is the between groups the smallest number and the within populations the biggest? the thing is that with so many loci, there are a lot that aren't really contributing to this differentiation between groups, they're just different between populations or whatever. and we also expect a fair amount of drift. \n",
    "\n",
    "the reason that the pvalue is all 0 for all the categories is that with Fst or FCT its testing whether the results are significantly different than zero, and basically any small number is going to be significantly different than zero. this doesnt tell us whether the amount or difference is biologically significant. \n",
    "\n",
    "for the AMOVA we want to report the overall FCT \n",
    "\n",
    "\n",
    "\n",
    "july 30 \n",
    "\n",
    "Im running arlequin to get the Fct to use that value to compare them. its not because there  is actually a way to run the program in a different hierarchical way. \n",
    "\n",
    "I made the individual files from the 16681_80_arleqin.txt file and just cut out the populations that i didnt want to compare, the same way as with the bayescan. i did the 7 within river comparisons. \n",
    "\n",
    "i put them in a folder called RIvers in the Arlequin directory, because they were getting all mixed up with the earlier runs, I felt like they were piggy backing off some of the premade program files. So all alone, I started tauy, i think its going to take a while?  I had to define the structure, so I put them as separate entities. I hope it works. \n",
    "\n",
    "it kindof does, it wants to run them but it give sme this error that says that the f statistics are outside of the range. garrett said that he had that problem before and that it has to do with the version of arlequin more than anything else, and he gave me an older version to try. Its one version earlier than the one that I have, and i put it in the folder where I was running things, on seagate. \n",
    "\n",
    "nope, it gives a different set of errors: \n",
    "\n",
    "==============================================================================\n",
    "== ANALYSES AT THE INTRA-POPULATION LEVEL\n",
    "==============================================================================\n",
    "==============================================================================\n",
    "== Sample : \tSNOH03\n",
    "==============================================================================\n",
    "Ewens-Watterson test cannot be computed \n",
    "in case of genotypic data with unknown gametic phase\n",
    "\n",
    "We hope to change it in a future release...\n",
    "===============================================================================\n",
    "== Sample : \tSNOH96\n",
    "===============================================================================\n",
    "Ewens-Watterson test cannot be computed \n",
    "in case of genotypic data with unknown gametic phase\n",
    "\n",
    "We hope to change it in a future release...\n",
    "================================================================================\n",
    "== Summary of computations done within populations\n",
    "================================================================================\n",
    "==============================================================================\n",
    "== DETECTION OF LOCI UNDER SELECTION FROM GENOME SCANS BASED ON F-STATISTICS\n",
    "==============================================================================\n",
    "\n",
    "i tried it with a different set of data, but that didnt help. The option doesnt even look like its been chosen, so thats super strange. \n",
    "\n",
    "i opened the nome one in the 3.5.1.3- its weird because it starts out as the 3.5.1.2 version and then switches to the 3.5.1.3 version as soon as you import the data into it. its like it has a memory of the version of the software built into the premade files. so if i open something that already exists, it reverts to the earlier program that was using it. so that is lame, im not sure how to uncouple that information.\n",
    "\n",
    "i dont know how to build the file? like maybe go to the original genepop and break it down from there? \n",
    "\n",
    "August 29 \n",
    "\n",
    "Im looking for the results of the Arlequin run, but i dont see anything that looks like output files.  And Im starting to remember that I couldnt ever really get the program to run? I need to look back at the project files I think. Anyway, I still dont understand how im going to do the outlier analyses, like what are the different comparisons? And how are they answering the questions that Im posing? \n",
    "\n",
    "But maybe I didn finish them and I cnat remember where they are? I feel like i remember them all being together on some html page? \n",
    "\n",
    "This is what i got when i tried to do the tauy ones. \n",
    "\n",
    "\tObserved F-statistics lead to invalid migration rates. Coalescent simulations cannot be performed.\n",
    "Perform a simple AMOVA to check that all F-statistics are within the interval ]0..1[\n",
    "\n",
    "\n",
    "so i clicked the do amova button but it didnt like that either: \n",
    "\n",
    " #[WARNING # 3] : invalid structure: only one population per group\n",
    "  An AMOVA analysis was selected and no Genetic structure\n",
    "  is defined, AMOVA task is skipped.\n",
    "\n",
    "  \n",
    "So i went in and deleted the structure, putting them both at a zero structure, but this is the error i got then:\n",
    "\n",
    " #[WARNING # 6] : no samples have been defined\n",
    " #[WARNING # 7] : no group(s) have been defined\n",
    " #[WARNING # 8] : invalid structure: only one population per group\n",
    "  An AMOVA analysis was selected and no Genetic structure\n",
    "  is defined, AMOVA task is skipped.\n",
    "\n",
    "ok, gave them both a structure of one, so theyre in the same group \n",
    "\n",
    " #[WARNING # 9] : no samples have been defined\n",
    " #[WARNING # 10] : no group(s) have been defined\n",
    " #[WARNING # 11] : invalid structure: only one population per group\n",
    "  An AMOVA analysis was selected and no Genetic structure\n",
    "  is defined, AMOVA task is skipped.\n",
    "\n",
    "no better. \n",
    "\n",
    "So I opened the project called 16681_80_named.arp and am running the amova and the pop tests in there, its really really slow going. Its only 1% done after like 30 mins. \n",
    "\n",
    "\n",
    "\n",
    "Distance method: Pairwise differences\n",
    "\n",
    "--------------------------\n",
    "AMOVA design and results :\n",
    "--------------------------\n",
    "\n",
    "</data>\n",
    "<Reference>Weir, B.S. and Cockerham, C.C. 1984.</Reference>\n",
    "<Reference>Excoffier, L., Smouse, P., and Quattro, J. 1992.</Reference>\n",
    "<Reference>Weir, B. S., 1996.</Reference>\n",
    "<data>\n",
    "----------------------------------------------------------------------\n",
    " Source of                  Sum of      Variance         Percentage\n",
    " variation      d.f.        squares     components       of variation\n",
    "----------------------------------------------------------------------\n",
    " Among\n",
    " groups           1      33395.308       78.21364 Va             5.72\n",
    "\n",
    " Among\n",
    " populations\n",
    " within\n",
    " groups          12      40660.388       39.58588 Vb             2.89\n",
    "\n",
    " Among\n",
    " individuals\n",
    " within\n",
    " populations    369     454329.161      -18.55211 Vc            -1.36\n",
    " Within\n",
    " individuals    383     485777.500     1268.34856 Vd            92.74\n",
    "----------------------------------------------------------------------\n",
    " Total          765    1014162.356     1367.59597\n",
    "----------------------------------------------------------------------\n",
    " Fixation Indices\n",
    "      FIS :     -0.01484\n",
    "      FSC :      0.03070\n",
    "      FCT :      0.05719\n",
    "      FIT :      0.07257\n",
    "----------------------------------------------------------------------\n",
    "\n",
    "Significance tests (10100 permutations)\n",
    "------------------\n",
    "\n",
    "Vd and FIT : P(rand. value &lt; obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00000\n",
    "                                  P-value =  0.00000+-0.00000\n",
    "\n",
    "Vc and FIS : P(rand. value > obs. value)  =  0.96030\n",
    "             P(rand. value = obs. value)  =  0.00000\n",
    "                                  P-value =  0.96030+-0.00175\n",
    "\n",
    "Vb and FSC : P(rand. value > obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00000\n",
    "                                  P-value =  0.00000+-0.00000\n",
    "\n",
    "Va and FCT : P(rand. value > obs. value)  =  0.00000\n",
    "             P(rand. value = obs. value)  =  0.00050\n",
    "                                  P-value =  0.00050+-0.00022\n",
    "\n",
    "</data>\n",
    "\n",
    " <A NAME=\"31_08_15at13_29_11_amova_POP_AMOVA_FIS\"></A> \n",
    "<data>\n",
    "\n",
    "\n",
    "-------------------------------------------------------\n",
    "Population specific FIS indices (10100 permutations)\n",
    "---------------------------------------------------\n",
    "  Pop#      Name       FIS    P(Rand FIS>=Obs FIS)\n",
    "---------------------------------------------------\n",
    "     1    AMUR10  -0.00541            0.603267\n",
    "     2   HAYLY10  -0.02467            0.801485\n",
    "     3   KOPPE96  -0.01611            0.633762\n",
    "     4   KUSHI06  -0.01680            0.710396\n",
    "     5    NOME94   0.01377            0.394653\n",
    "     6    SNOH96  -0.02197            0.730099\n",
    "     7    TAUY12  -0.02199            0.782079\n",
    "     8    AMUR11  -0.02647            0.840099\n",
    "     9   HAYLY09  -0.02414            0.798812\n",
    "    10   KOPPE91   0.00188            0.468911\n",
    "    11   KUSHI07  -0.00656            0.599307\n",
    "    12    NOME91  -0.00821            0.625743\n",
    "    13    SNOH03  -0.00919            0.619505\n",
    "    14    TAUY09  -0.02811            0.845446\n",
    "---------------------------------------------------\n",
    "\n",
    "\n",
    "********************\n",
    "Locus by locus AMOVA:\n",
    "********************\n",
    "\n",
    "Distance method for locus-by-locus analysis: Pairwise differences\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "I am still trying to figure out arlequin. Im pretty sure I never got it to work. On July 23 i changed to adegente to do the amova at least, though i never got adegenet to work either. but for that program  i made the file 16681_80_edit.gen which is a file that i was looking at and wondering about earlier. \n",
    "\n",
    "         $$ September 1 2015\n",
    "==================================================================\n",
    "Still trying to get Arlequin to run, it crashed overnight. Garrett says he had 9 pops and about 10 thousand loci and it ran, so i might just be topping out what my computer can do, or the program can handle. \n",
    "\n",
    "I started the Asia file that I made yesterday, it has the asian samples, no nome and splits them up into even and odd.\n",
    "\n",
    "\n",
    "\t\t\t \n",
    "I moved over to Garretts computer to run the Arlequin and its kindof working. It did some of the tests on the 16681_80_named file, but it failed when it came to the outlier analysis because it says that the migration rate is something. So Garrett says to look over and see that none of the loci are fixed, that might be what is going on. There cant be any fixed loci and if the loci were filtered first and then there were individuals removed, it could change some of the loci, so they passed the minor allele filters when there were more individuals with bad genotyping but they wouldnt pass now with those loci removed. \n",
    "\n",
    "im not sure how to go about that, so i started the Asian file running, just an amova between the odd and the even and an outlier test. \n",
    "\n",
    "the detect loci under selection uses a hierarchical model and 20,000 simulations and 100 demes, 50 groups and an He min or 0 and an He max of 0.7 and the distance method. i thnk that those should be the defaults for the most part. \n",
    "\n",
    "Ok so the same thing happened, it ran the amova and did some indices, but it wont do the evans-williams test or whatever (some test for neutrality) because the data isnt phased, not that i asked for it to do that test. and it wont do the actual outlier analysis because it:\n",
    "\n",
    "\tObserved F-statistics lead to invalid migration rates. Coalescent simulations cannot be performed.\n",
    "Perform a simple AMOVA to check that all F-statistics are within the interval ]0..1[\n",
    "\n",
    "\n",
    "here is what it ouputs, I think that i need to trim these from the file to get it to run: \n",
    "\n",
    "Locus by locus AMOVA:\n",
    "********************\n",
    "\n",
    "Distance method for locus-by-locus analysis: Pairwise differences\n",
    "\n",
    "List of loci with only one allele:\n",
    "----------------------------------\n",
    "\n",
    "13 18 60 65 106 211 224 225 310 327 337 425 454 459 485 562 577 596 597 605 607 623 628 637 678 768 770 779 795 813 829 916 929 978 1019 1030 1034 1038 1054 1080 1091 1100 1153 1177 1179 1183 1237 1398 1401 1408 1462 1482 1496 1512 1522 1540 1564 1576 1605 1656 1714 1741 1746 1855 1886 1906 1930 1986 2016 2048 2084 2091 2096 2101 2110 2136 2153 2161 2188 2191 2256 2264 2284 2289 2326 2359 2387 2393 2420 2429 2432 2478 2547 2555 2561 2587 2603 2639 2645 2652 2653 2659 2712 2733 2757 2812 2854 2967 2986 3001 3019 3038 3054 3112 3118 3131 3146 3206 3270 3295 3316 3349 3402 3461 3469 3471 3508 3540 3576 3621 3715 3793 3794 3884 3932 3957 3986 4021 4032 4052 4068 4077 4112 4137 4156 4174 4193 4217 4246 4282 4298 4310 4311 4341 4378 4379 4388 4420 4447 4483 4536 4539 4544 4562 4567 4598 4610 4620 4630 4654 4665 4682 4697 4720 4743 4755 4767 4772 4782 4806 4836 4838 4894 4951 4961 4967 4977 5059 5089 5097 5113 5218 5225 5251 5280 5289 5352 5419 5430 5511 5539 5552 5585 5812 5830 5847 5855 5868 5893 5910 5981 6022 6054 6133 6273 6278 6343 6354 6358 6388 6393 6417 6445 6482 6503 6533 6600 6618 6637 6649 6654 6657 6659 6693 6702 6770 6811 6859 6908 6910 6966 6993 7011 7014 7028 7046 7050 7056 7065 7075 7077 7141 7143 7155 7182 7198 7215 7216 7241 7259 7291 7364 7422 7446 7479 7518 7541 7564 7583 7601 7623 7630 7639 7687 7694 7758 7809 7830 7844 7855 7932 7941 7985 7997 8004 8019 8028 8037 8041 8057 8140 8176 8187 8218 8226 8233 8239 8261 8267 8350 8363 8387 8438 8440 8451 8490 8527 8614 8624 8650 8658 8691 8711 8720 8745 8758 8762 8771 8797 8801 8803 8818 8819 8820 8826 8854 8856 8882 8912 8933 8940 8945 8958 8965 8989 9049 9059 9081 9151 9156 9160 9195 9207 9222 9296 9301 9306 9310 9336 9359 9395 9430 9443 9455 9460 9495 9505 9520 9640 9652 9750 9763 9768 9791 9829 9852 9855 9857 9881 9890 9893 9900 9918 9968 10020 10026 10060 10090 10114 10121 10138 10141 10155 10168 10196 10204 10232 10257 10297 10324 10344 10389 10405 10433 10437 10451 10455 10458 10480 10482 10499 10501 10502 10507 10526 10534 10590 10600 10623 10653 10654 10658 10659 10672 10678 10688 10695 10696 10715 10719 10726 10735 10758 10763 10766 10784 10793 10803 10843 10940 10942 10956 10979 11002 11038 11041 11042 11054 11070 11081 11189 11198 11255 11296 11303 11358 11366 11367 11374 11390 11395 11402 11405 11485 11487 11517 11568 11626 11631 11641 11657 11660 11670 11677 11689 11694 11702 11752 11761 11763 11766 11833 11834 11851 11857 11880 11881 11901 11922 11985 12068 12077 12106 12111 12132 12143 12151 12178 12182 12185 12238 12252 12266 12282 12285 12291 12394 12443 12463 12482 12486 12520 12538 12550 12583 12603 12633 12658 12681 12683 12694 12726 12734 12771 12778 12827 12834 12847 12884 12942 12954 12977 13000 13003 13004 13022 13074 13105 13109 13110 13117 13181 13214 13228 13251 13310 13383 13385 13404 13451 13463 13499 13512 13525 13533 13535 13554 13570 13584 13587 13592 13612 13614 13631 13641 13687 13696 13700 13703 13723 13725 13777 13788 13791 13792 13816 13868 13877 13899 13925 13942 13980 13982 13995 14041 14065 14081 14088 14099 14106 14179 14183 14195 14210 14222 14247 14254 14260 14270 14271 14277 14420 14440 14450 14460 14469 14470 14473 14492 14495 14502 14516 14521 14539 14554 14580 14609 14637 14638 14658 14666 14694 14697 14715 14728 14732 14747 14752 14759 14760 14772 14788 14789 14810 14818 14831 14835 14846 14868 14871 14872 14874 14902 14929 14930 14949 14957 14971 14993 15025 15027 15043 15045 15054 15059 15106 15110 15151 15167 15175 15181 15198 15220 15222 15251 15266 15274 15292 15304 15325 15334 15369 15374 15383 15398 15406 15409 15423 15433 15475 15486 15502 15512 15517 15518 15521 15526 15532 15539 15546 15550 15557 15621 15626 15662 15699 15712 15755 15781 15812 15838 15840 15872 15914 15915 15944 15954 15961 15977 16016 16030 16044 16063 16074 16079 16094 16097 16131 16207 16210 16212 16315 16317 16364 16376 16394 16395 16408 16417 16425 16430 16443 16463 16492 16512 16522 16528 16543 16545 16555 16572 16623 16646 16660 16666 16674 \n",
    "\n",
    "\n",
    "\n",
    "when the program kicks these out on its own, there are 15945 loci left that are polymorphic. Im not sure how Im going to go about removing them from the input file? Is there any way to easily build a script that does it? Or should I go back to the genepop file and remake the input? That might be the easiest way to go about it. \n",
    "\n",
    "Ok, so i took them into exce, and took the list of the loci names from the genepop file 16681_80_genepop_one_per_line and used a couple vlookups to get a list of the monomorphic actual radtag names (theyre just the index right now) and a list of the polymorphic loci. I used the perl script to subset the genepop files to just keep the polymorphic ones- and called it 15945_polymorphic.txt\n",
    "\n",
    "\n",
    "G:\\Analysis\\Pop_analysis\\Populations_b3_may\\Arlequin>perl subset_genepop_by_SNPs.pl polymorphic.txt 16681_80_genepop_one_per_line.txt >15945_polymorphic.txt\n",
    "\n",
    "I think i converted that using PGDspider last time....\n",
    "\n",
    "INFO  14:41:11 - Saved SPID file as: genepop_arlequin2.spid\n",
    "INFO  14:41:11 - loading SPID file: genepop_arlequin2.spid\n",
    "INFO  14:46:02 - Convert...\n",
    "INFO  14:46:02 - convert GENEPOP:15945_polymorphic.txt to Arlequin:15945_polymorphic.arp\n",
    "WARN  14:48:23 - The name of a population was missing (no name attribute). It is now recoded as \"pop_1\"\n",
    "WARN  14:48:25 - The name of a population was missing (no name attribute). It is now recoded as \"pop_2\"\n",
    "WARN  14:48:27 - The name of a population was missing (no name attribute). It is now recoded as \"pop_3\"\n",
    "WARN  14:48:29 - The name of a population was missing (no name attribute). It is now recoded as \"pop_4\"\n",
    "WARN  14:48:31 - The name of a population was missing (no name attribute). It is now recoded as \"pop_5\"\n",
    "WARN  14:48:32 - The name of a population was missing (no name attribute). It is now recoded as \"pop_6\"\n",
    "WARN  14:48:34 - The name of a population was missing (no name attribute). It is now recoded as \"pop_7\"\n",
    "WARN  14:48:36 - The name of a population was missing (no name attribute). It is now recoded as \"pop_8\"\n",
    "WARN  14:48:38 - The name of a population was missing (no name attribute). It is now recoded as \"pop_9\"\n",
    "WARN  14:48:39 - The name of a population was missing (no name attribute). It is now recoded as \"pop_10\"\n",
    "WARN  14:48:40 - The name of a population was missing (no name attribute). It is now recoded as \"pop_11\"\n",
    "WARN  14:48:42 - The name of a population was missing (no name attribute). It is now recoded as \"pop_12\"\n",
    "WARN  14:48:43 - The name of a population was missing (no name attribute). It is now recoded as \"pop_13\"\n",
    "WARN  14:48:45 - The name of a population was missing (no name attribute). It is now recoded as \"pop_14\"\n",
    "INFO  14:48:49 - translation completed\n",
    "\n",
    "I opened it and it looks fine, I added the names of the populations and saved it. \n",
    "\n",
    "i want to edit again into the continents. I did two asian, one with nome and one without and a north american. They dont have their structure in there yet. \n",
    "\n",
    "\n",
    "I cant get Arlequin to open, I think it could be because I just have too much running. I am at 100% of my CPU with the lepmap and R both going. and they would all take more if they could. \n",
    "\n",
    "i got the Asian one to load and split it up odd and even and started an amova and the detect loci under selection. Maybe this time it will work because there are only polymorphic loci? \n",
    "\n",
    "        $$ September 2 2015\n",
    "==================================================================\n",
    "\n",
    "\n",
    "> pairwise.fst(pops_16681_edit, pop= NULL, res.type = c(\"dist\", \"matrix\"))\n",
    "            1          2          3          4          5          6          7          8          9         10         11         12         13\n",
    "2  0.03713108                                                                                                                                    \n",
    "3  0.03578918 0.01066476                                                                                                                         \n",
    "4  0.01221767 0.03705546 0.03560100                                                                                                              \n",
    "5  0.05044387 0.02753836 0.02758697 0.04996322                                                                                                   \n",
    "6  0.02373888 0.04728466 0.04579836 0.02243125 0.05672948                                                                                        \n",
    "7  0.01378259 0.04147544 0.04081558 0.01926097 0.05490709 0.03093015                                                                             \n",
    "8  0.04442862 0.01364543 0.01718015 0.04505148 0.03260416 0.05492264 0.04877038                                                                  \n",
    "9  0.03663895 0.01211269 0.01108887 0.03654334 0.02965242 0.04874063 0.04172586 0.01863162                                                       \n",
    "10 0.01681382 0.03823536 0.03730709 0.01340815 0.05482867 0.02718595 0.02274181 0.04622084 0.04082480                                            \n",
    "11 0.06849168 0.04433646 0.04540019 0.06731200 0.03558518 0.07585488 0.07109011 0.04972008 0.04760146 0.07065036                                 \n",
    "12 0.03469519 0.05091818 0.05060942 0.03275280 0.05994755 0.02813025 0.03828925 0.05899044 0.05192792 0.03442890 0.07182033                      \n",
    "13 0.03591335 0.01107856 0.01039770 0.03568867 0.02850251 0.04682780 0.04060512 0.01754976 0.01205706 0.03832820 0.04595166 0.05020727           \n",
    "14 0.01208570 0.03693895 0.03532381 0.01046837 0.05122340 0.02395459 0.01880507 0.04483438 0.03694565 0.01481165 0.06871101 0.03375420 0.03551611\n",
    "\n",
    "\n",
    "\n",
    "I copied over the edited arlequin files to my work drive and tried running the asian one on garretts computer, but got the same error about the FST being out of range. So I dont really have any other ideas about what to do. I think that Arlequin has taken up enough of my time at this point I need to move over to programs that are working. \n",
    "\n",
    "Sep 4 \n",
    "\n",
    "\n",
    "In the Arlequin file there are some results: \n",
    "\t16681_80_Asian \n",
    "\t\tAmova: even vs odd\n",
    "\t\t\tDistance method: Pairwise differences\n",
    "\t\t\t\n",
    "\t15945_Asian \n",
    "\t\tAmova: even vs odd; the \"polymorphic\" loci \n",
    "\t\t\tDistance method: Pairwise differences\t\t\t\n",
    "\t\n",
    "\t16681_80_named \n",
    "\t\tAmova: even vs odd; all pops\n",
    "\t\t\tDistance method: Pairwise differences\n",
    "\t\t\t\n",
    "The 15945 are the loci that were identified as polymorphic from the second 16681 run of the asian samples in arlequin and so i took them out and made the other file, its probably not appropriate to run these loci in the NA samples... ? \n",
    "\n",
    "\n",
    "\n",
    "I restarted 16881_NA in Arlequin on Garretts computer, an AMOVA with the Fis, so that i could have it, I also asked it to do the outlier analysis, but it probably wont. \n",
    "\n",
    "It gave me a list of the loci that are monomorphic in the NA pops: \n",
    "********************\n",
    "Locus by locus AMOVA:\n",
    "********************\n",
    "\n",
    "Distance method for locus-by-locus analysis: Pairwise differences\n",
    "\n",
    "List of loci with only one allele:\n",
    "----------------------------------\n",
    "\n",
    "173 230 252 270 299 313 535 536 585 652 703 759 842 930 977 1012 1020 1041 1070 1141 1184 1235 1236 1298 1383 1391 1499 1527 1691 1725 1735 1824 1842 1847 1934 1995 2024 2038 2080 2115 2165 2186 2255 2260 2275 2288 2386 2614 2641 2663 2664 2678 2697 2794 2816 2848 2922 2942 3166 3234 3250 3265 3300 3315 3421 3429 3608 3744 3756 3892 3914 4004 4011 4014 4146 4162 4202 4281 4326 4382 4414 4451 4454 4457 4990 4995 5248 5492 5648 5724 5729 5952 5972 6064 6070 6155 6190 6197 6222 6350 6430 6470 6494 6631 6651 6656 6663 6720 6745 6798 6820 6867 6920 7003 7104 7202 7225 7303 7396 7404 7426 7456 7650 7787 7912 7935 7966 8058 8129 8174 8377 8402 8423 8436 8585 8594 8652 8662 8714 8747 9086 9194 9295 9300 9346 9362 9529 9543 9624 9734 9799 9821 9868 9878 9960 10008 10147 10181 10444 10467 10567 10711 10724 10757 10913 10946 10982 10992 11052 11056 11061 11151 11196 11225 11259 11387 11478 11531 11623 11633 11643 11699 11794 12169 12213 12322 12353 12362 12363 12448 12472 12602 12626 12640 12648 12670 12908 13010 13055 13077 13239 13290 13312 13450 13469 13521 13567 13603 13605 14016 14069 14133 14234 14344 14372 14439 14444 14511 14519 14572 14574 14600 14602 14632 14680 14751 14758 14820 15061 15090 15196 15206 15275 15290 15310 15359 15415 15425 15499 15551 15556 15592 15597 15641 15653 15658 15670 15725 15737 15852 15916 15934 16051 16078 16090 16133 16157 16165 16176 16267 16277 16292 16324 16379 16411 16455 16477 16539 16581 16636 16641 16663 \n",
    "\n",
    "remade the file for NA to see if I can get it to do the outlier test: \n",
    "\n",
    "perl subset_genepop_by_SNPs.pl NApolymorphic.txt 16681_80_genepop_one_per_line.txt >16409_NApolymorphic.txt\n",
    "\n",
    "INFO  15:48:03 - loading SPID file: genepop_arlequin2.spid\n",
    "INFO  15:48:13 - SPID file was not saved.\n",
    "INFO  15:48:17 - Convert...\n",
    "INFO  15:48:17 - convert GENEPOP:16409_NApolymorphic.txt to Arlequin:16609_NA_poly.arp\n",
    "WARN  15:51:30 - The name of a population was missing (no name attribute). It is now recoded as \"pop_1\"\n",
    "WARN  15:51:32 - The name of a population was missing (no name attribute). It is now recoded as \"pop_2\"\n",
    "WARN  15:51:35 - The name of a population was missing (no name attribute). It is now recoded as \"pop_3\"\n",
    "WARN  15:51:37 - The name of a population was missing (no name attribute). It is now recoded as \"pop_4\"\n",
    "WARN  15:51:40 - The name of a population was missing (no name attribute). It is now recoded as \"pop_5\"\n",
    "WARN  15:51:42 - The name of a population was missing (no name attribute). It is now recoded as \"pop_6\"\n",
    "WARN  15:51:43 - The name of a population was missing (no name attribute). It is now recoded as \"pop_7\"\n",
    "WARN  15:51:46 - The name of a population was missing (no name attribute). It is now recoded as \"pop_8\"\n",
    "WARN  15:51:48 - The name of a population was missing (no name attribute). It is now recoded as \"pop_9\"\n",
    "WARN  15:51:50 - The name of a population was missing (no name attribute). It is now recoded as \"pop_10\"\n",
    "WARN  15:51:52 - The name of a population was missing (no name attribute). It is now recoded as \"pop_11\"\n",
    "WARN  15:51:54 - The name of a population was missing (no name attribute). It is now recoded as \"pop_12\"\n",
    "WARN  15:51:56 - The name of a population was missing (no name attribute). It is now recoded as \"pop_13\"\n",
    "WARN  15:51:59 - The name of a population was missing (no name attribute). It is now recoded as \"pop_14\"\n",
    "INFO  15:52:06 - translation completed\n",
    "\n",
    "\n",
    "I changed the names and cut out the Asian populations and started it running on GArretts computer. \n",
    "\n",
    "\n",
    "         $$ September 9 2015\n",
    "==================================================================\n",
    "\n",
    "I got distracted thinking of the Amovas that I should run in Arlequin for the NA and Asian samples, \n",
    "\n",
    "I need to do what lisa did and do two more hierarchical Amovas, both that are the rivers grouped together irregardless of the lineage. so amur and amur vs haylyl and halyl vs tauy and tauy etc. \n",
    "\n",
    "i started with NA and used the 16409 set that are polymorphic. UGh i forget that every time i try to run arlequin it quits in the middle of the night. I might be better off just quitting this and starting it on Garretts computer. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#making the input files \n",
    "\n",
    "\n",
    "I looked at the arlequin input for the AMOVA etc and there are 32816 genotypes? in the first sample that came up for the (misslabled- i changed the names for all the input and the output files to 16409) 16609_NA_named_poly.arp file, which if you devide by 2 is 16408 :( so i guess we assume that the last locus was left of that file too? And there are only 16408 loci in the outputs, and it says that there are only 16408 total loci at the top of the results file. \n",
    "\n",
    "But for the arlequin results for the 16681 file there are 16681 actual loci represented... so i dont know what to say. \n",
    "\n",
    "\n",
    "In the mean time, Arlequin crashed. It said there was an EAccessViolation. Im going to restart them on Garretts later, Ill do both the Asian and the North American, even though there is this problem. Im going to go forward and just probably throw that locus out later.  \n",
    "\n",
    "\n",
    "I made a new file in the Arlequin file that is for the Summary stats of these arlequin data sets (and they might be used for the other analyses too-) its called SUMStats and i put in the original genepop file for 16681 and the 16409 and the 15945 and Ill run the several perl and python scripts that I have to do summary stats on them. Count Hets, Count missing, cout allele freq etc. \n",
    "\t\n",
    "Using a command from July: \n",
    "\n",
    "perl countHets.pl 15945_polymorphic.txt >countHets_15945.txt\n",
    "perl countMissing.pl 16681_80_genepop_one_per_line.txt >countMissing_16681_80.txt\n",
    "\n",
    "The several things that I have tried to get this perl script to run on these samples have not worked. So I wonder? But there are no isntructions in the script and its written in perl which i dont have a clue about. \n",
    "\n",
    "And those commands on that file dont even work. so awesome. \n",
    "\n",
    "I emailed garrett to get his help on it, and tomorrow I want to make sure that I have the files that I want to use for the remainder of the analysis. \n",
    "\n",
    "and i really want to finish inventorying the analysis that i have done, and maybe come up with a plan for how to finish the rest of what i need to do. \n",
    "  $$ September 11 2015\n",
    "==================================================================\n",
    "\n",
    "Garretts scripts have no information associated with them as to how to use them or what type of file they work on, so it turns out that those were for the output from stacks, and they are named the same as the scripts that are for the output of genepop. ITs just a different countHets.pl that I needed. \n",
    "\n",
    "\n",
    "perl countHets.pl 15945_polymorphic.txt >countHets_15945.txt\n",
    "perl countHets.pl 16409_NApolymorphic.txt >countHets_16409_NA.txt\n",
    "\n",
    "perl countMissing.pl 15945_polymorphic.txt >countMissing_15945.txt\n",
    "perl countMissing.pl 16409_NApolymorphic.txt >countMissing_16409_NA.txt\n",
    "\n",
    "\n",
    "I wanted to make sure that the 16409 file was right, so I took 100029_14 out of the list of polymorphic loci, and reran the subset_genepop perl script on the 16881 file and what came out was totally normal looking. I named it 16408_polymorphic.txt and im going to go forward with the analysis using that file. \n",
    "\n",
    "Ran the stats on that file: \n",
    "\n",
    "perl countHets.pl 16408_polymorphic.txt >countHets_16408.txt\n",
    "perl countMissing.pl 16408_polymorphic.txt >countMissing_16408.txt\n",
    "\n",
    "Then put them both into geneopop and got the allele frequencies per allele and individual, and took the .INF output file and ran through wes' script that does some counts: \n",
    "\n",
    "\n",
    "perl genepop_al_freqs+al_counts.pl 16408_polymorphic.txt.inf POPnames.txt > genepop_al_freq_counts_16408.txt\n",
    "\n",
    "perl genepop_al_freqs+al_counts.pl 15945_polymorphic.txt.inf POPnames.txt > genepop_al_freq_counts_15945.txt\n",
    "\n",
    "\n",
    "I put all the results of all in an excel sheet, and Im just looking for all the things we filtered for before. I found something right away, the 16408 file is missing an individual: PKUSHI06_0006_comb\n",
    "and when I go to the 16681_80_genepop_one_per_line file that was used to make it I see that the Kushiro pop is actually pretty scrambled and the entire file has some strange inconsistencies. Im worried for any input file that got made with this. \n",
    "\n",
    "So I deleted all the files for 16408 and 16409 and 15945 from this run and the 16681_genepop file that is wrong and start looking through all the files for PKUSHI06_0006_comb in the outputs, and the inventory turns up that Arlequin basically all needs to be remade as does the Bayescan file, and from there I can check to see if the new bayescan file matches the inputsf for Bayescan, BayesEnv and BayeScEnv. Those have a file format that has no individual information, so I cant tell what is wrong and what isnt. I just have to make a new one that I know is right to compare to maek sure i had the correct input files when I started running them. \n",
    "\n",
    "I deleted a bunch of things, but I also forgot or didnt realize that I had run the BayeScEnv on another computer, probably Ryans, a while ago and it was on olympus, so I moved that over and in general just deleted and cleaned up things that I didnt think i would ever need. \n",
    "\n",
    "OK Rebuilding the Arlequin files: \n",
    "\n",
    "perl genepop_print_snp_names_1_per_line.pl batch_3_16681_pop_80.genepop > batch_3_16681_pop_80_one_line.genepop \n",
    "That didnt actually work, I went in and did the names of the loci by hand, replacing the commas with an \\n\n",
    "\n",
    "I looked at how many genotypes were in the first individual, and there were 16681, just as many as we would expect, because 100029_14 is still in this file and im going to roll with it and see if it works? \n",
    "\n",
    "perl subset_genepop_by_SNPs.pl NA.txt batch_3_16681_pop_80_one_line.genepop > 16409_polymorphic.txt\n",
    "perl subset_genepop_by_SNPs.pl ASIAN.txt batch_3_16681_pop_80_one_line.genepop > 15945_polymorphic.txt\n",
    "\n",
    "Now Im putting the three files through pgdspider so that they can be translated into the GESTE format. \n",
    "\n",
    "\n",
    "INFO  13:48:09 - Convert...\n",
    "INFO  13:48:09 - convert GENEPOP:15945_polymorphic.txt to GESTE / BayeScan:15945_polymorphic_bayescan.txt\n",
    "INFO  13:53:37 - translation completed\n",
    "INFO  13:56:52 - Convert...\n",
    "INFO  13:56:52 - convert GENEPOP:batch_3_16681_pop_80_one_line.genepop.txt to GESTE / BayeScan:batch_3_16681_pop_80.bayescan.txt\n",
    "INFO  14:01:32 - translation completed\n",
    "INFO  14:03:05 - Convert...\n",
    "INFO  14:03:05 - convert GENEPOP:16409_polymorphic_genepop.txt to GESTE / BayeScan:16409_polymorphic_bayescan.txt\n",
    "\n",
    "compare this file to the ones that were used to run the bayes programs Bayescan and BayeScEnv\n",
    "\n",
    "this bayescan file is the same (in a spot check) as the one that was used for the bayescan runs. That means that the files are the same for the BayeScEnv runs as well \n",
    "\n",
    "The Arlequin file for 16881 looks the same, so I moved all those runs back into the folder, but I still want to redo the 15945 and the 16409 ones. \n",
    "\n",
    "Then manually edit to remove the spaces, turn them into tabs\n",
    "\n",
    "then use the python script that i wrote to convert that file into bayesenv2 and check to make sure that it is the same as those. (though if it was the same as the other Bayes programs it probably has already passed this test) \n",
    "\n",
    "perl countHets_genepop.pl 15945_polymorphic_genepop.txt >countHets_15945.txt\n",
    "perl countHets_genepop.pl 16409_polymorphic_genepop.txt >countHets_16409_NA.txt\n",
    "\n",
    "perl countMissing_genepop.pl 15945_polymorphic_genepop.txt >countMissing_15945.txt\n",
    "perl countMissing_genepop.pl 16409_polymorphic_genepop.txt >countMissing_16409_NA.txt\n",
    "\n",
    "\n",
    "Redoing the genepop analysis on these two files, doing the option 5,1 to get the INF file, to get the allele counts and the allele frequencies. \n",
    "\n",
    "perl genepop_al_freqs+al_counts.pl 16409_polymorphic_genepop.txt.inf POPnames.txt > genepop_al_freq_counts_16409.txt\n",
    "\n",
    "perl genepop_al_freqs+al_counts.pl 15945_polymorphic_genepop.txt.inf POPnames.txt > genepop_al_freq_counts_15945.txt\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genepop "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have a Vcf file for the 80%, i made it on the 26th, and I will edit a genepop file to also be the set of 80%. Ill make a new file that is the final set of data- call it the 16681and80 or something, and put those files in it. From there I can make the PCAs. I need to  make a checklist of all the filters. Right now I have the traditional ones, but I haven't looked at allelic bias and i havent done anything with the number of alleles.    \n",
    "\n",
    "i deleted the individuals that didnt meet the threshold in the genepop folder and renamed the folder 16681_80_genepop.txt\n",
    "\n",
    "\n",
    "Genepop version 4.3\n",
    "\n",
    "Name of the data file ? (press ENTER to quit)\n",
    "16681_80_genepop_one_per_line.txt\n",
    "Current input file: 16681_80_genepop_one_per_line.txt\n",
    "(!) Invalid information for some individual: some information was found\n",
    "   after all genotypes have been read for an individual.\n",
    "   Check individual 1 in population 1   Suspect information: '0303'\n",
    "\n",
    "I used the batch3_16681_oneperline.txt file that i found on Seagate in populations_b3_may and brought that over to the windows side, it looks right, so I added the population delimination and deleted the individuals that didnt make the 80% threshold. Its loading into geneopop, run out of the Ubuntu_shared folder. \n",
    "\n",
    "\n",
    "File: batch3_16681_oneperline_80.txt (Title Line:)\n",
    "\n",
    "Number of populations detected : 14\n",
    "Number of loci detected        : 16681\n",
    "\n",
    "Indices for populations:\n",
    "----     -------------\n",
    "1        PAMUR10_\n",
    "2        PAMUR11_\n",
    "3        PHAYLY09\n",
    "4        PHAYLY10\n",
    "5        PKOPE91T\n",
    "6        PKOPE96T\n",
    "7        PKUSHI06\n",
    "8        PKUSHI07\n",
    "9        PNOME91_\n",
    "10       PNOME94_\n",
    "11       PSNOH03_\n",
    "12       PSNOH96_\n",
    "13       PTAUY09_\n",
    "14       PTAUY12_\n",
    "----------------------\n",
    "\n",
    "Estimates for all loci (diploid):\n",
    "=========================\n",
    "pop      1       2       3       4       5       6       7       8       9       10      11      12      13      \n",
    "2      0.0634 \n",
    "3      0.0594  0.0029 \n",
    "4      0.0062  0.0648  0.0601 \n",
    "5      0.0886  0.0392  0.0385  0.0889 \n",
    "6      0.0303  0.0828  0.0780  0.0269  0.0939 \n",
    "7      0.0101  0.0760  0.0727  0.0231  0.1015  0.0465 \n",
    "8      0.0767  0.0095  0.0177  0.0791  0.0490  0.0956  0.0883 \n",
    "9      0.0588  0.0029  0.0008  0.0596  0.0382  0.0781  0.0728  0.0181 \n",
    "10     0.0101  0.0679  0.0627  0.0020  0.0923  0.0282  0.0278  0.0819  0.0623 \n",
    "11     0.1295  0.0800  0.0801  0.1305  0.0519  0.1359  0.1414  0.0876  0.0805  0.1341 \n",
    "12     0.0537  0.1003  0.0957  0.0520  0.1105  0.0295  0.0684  0.1116  0.0952  0.0537  0.1479 \n",
    "13     0.0598  0.0026  0.0009  0.0609  0.0391  0.0792  0.0735  0.0171  0.0013  0.0642  0.0813  0.0968 \n",
    "14     0.0049  0.0642  0.0589  0.0008  0.0891  0.0288  0.0213  0.0783  0.0586  0.0035  0.1306  0.0530  0.0599 \n",
    "\n",
    "The file batch3_16681_oneperline_80.txt.MIG contains the matrix ready for further analysis\n",
    "Normal ending.\n",
    "\n",
    "\n",
    "\n",
    "im running genepop on that file of 16681 and the 80 and trying to get some of these metrics calculated. ALL OF THIS IS BEING RUN IN UBUNTU SHARED because the external is still on the linux side of the machine. I did 6,2 (pariwise Fst, .ST2) and 5,2 (gene diversities and Fis per locus and per sample .DIV) and 5,1 (allele and genotype frequencies per locus and per sample, .INF) \n",
    "\n",
    "\n",
    "Sep 14\n",
    "\n",
    "\n",
    "Running genepop on the 15945_polymorphic_genepop.txt and 16409_polymorphic_genepop.txt. I also cleaned up the files, moving them out of Arlequin, and organizing them with in a file into their own analysis, considering that they have bayescan and genepop files. \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
